# 03_Feature Engineering — GBDT 共通 → GBDT モデル別 → RealMLP

4モデル(LightGBM / XGBoost / CatBoost / RealMLP)が本番で使っている特徴量を、**上から順番に実行する**ノートブック。

各節では特徴量の定義、導入理由等を記載している。

---

## 0.このノートブックの流れ

| 章 | 内容 | 対象モデル |
|---|---|---|
| 1 | **EDA の仮説と、この工程での検証内容**(何を確かめ、どうなったか) | — |
| 2 | データ前処理(読み込み・fold 分割) | — |
| 3 | **GBDT 3種で共通**の Feature Engineering: 厳密値キー / Smooth Keys / 作るエンコーディングの設計 / Target Encoding / digit / Count /(補足)矛盾しないか | LightGBM・XGBoost・CatBoost |
| 4 | **GBDT のモデル別**の Feature Engineering と、組み上がった列が本番と一致するかの確認。4-4 に列の一覧 | 各モデル |
| 5 | **RealMLP** の Feature Engineering(GBDT とは別系統) | RealMLP |
| 6 | スコア(04 の実行結果のメモ) | — |
| 7 | `.py` と同じ結果になるかの確認 　**※コードの実行はここまで** | — |
| 8 | (補足)`.py` との構成の違い | — |
| 9 | (補足)不採用にした施策(EDA の仮説から検証したものも含む) | — |

- **`src/` の `.py` は読み込まず、`read_csv` から特徴量の作成までをこのノートブックの中だけで行う。**
  本番(`src/03_feature_engineering_<model>.py`)と同じ結果になることは 7 章で値まで比較して確かめる。
- モデル学習・評価はここでは行わず、 `src/04_train_and_evaluate_<model>.py` で定義・実行する(6章)。
- 改善幅は OOF AUC の差。採用基準は **+0.00008 以上かつ paired DeLong 検定で z ≥ 3**

---

## 1. EDA の仮説と、この工程での検証

`01_eda.ipynb` の結論で、次工程に3つの施策(施策1〜3)を渡した。

加えて、EDA の所見から「列を減らす」「補助金を他の列と組み合わせる」ことも検証した(施策の列が空欄の行)。

下記の表は実施した施策及び 2. 以降のフローで検証した結果である。

| 仮説(EDA) | 施策(EDA の結論) | 検証した内容 | 結果(OOF AUC の差) | 判定 | 検証した工程・節 |
|---|---|---|---|---|---|
| 仮説1: 補助金・航続距離不安などカテゴリ列も強く効く | 施策1: カテゴリ列を落とさずモデルに渡す | カテゴリ6列をモデルに渡す | 0.8838 → 0.9412(**+0.058**) | **採用** | `02_baseline` |
| 仮説2: 年収は細かい値そのものに情報がある | 施策2: 値を丸めずに購入率を渡す、ビン数を引き上げる | 厳密値キーで Target Encoding | **+0.00345**(LightGBM)/ +0.00108(CatBoost) | **採用** | 3-1 / 3-4 |
| 同上 | 同上 | 丸めずに見せる別の経路: ビン数 1024 + digit features + Smooth Keys + 平滑化3種の Target Encoding | +0.00052 / +0.00048 / **+0.00096**(LightGBM / XGBoost / CatBoost) | **採用** | 3-2 / 3-4 / 3-5 |
| 仮説4・6: 交互作用は「自宅充電 × スタンド数」の1組だけ | 施策3: 交互作用の特徴量は、作る前に本当に必要かを測る | 組み合わせの特徴量を1列足す | LightGBM +0.000012 / RealMLP +0.000028 | **採用**(LightGBM・RealMLP。採用基準には届かないが CV 最高の構成) | 3-3 / 5-8 |
| 仮説4: `City_Type` はスタンド数と同じ動きをする | | `City_Type` を削除 | -0.000017(z=-1.30) | 誤差 → 残す | 9-1 |
| 仮説1・3: 目的変数と相関が薄い年齢・性別・保有台数は効かない | | この3列を削除 | **-0.00044(z=-14.19)** | **悪化** | 9-1 |
| 仮説5: 補助金は誰にでも一律に効く | | 補助金との組み合わせ(Target Encoding・積) | -0.000016 / +0.000009 | 誤差 → 単体のまま | 9-1 |

**効いたのは「元の情報を失わずにモデルへ渡す」施策(施策1・2)。**「組み合わせる」「減らす」方向は、誤差か悪化だった。

---

## 2. データ前処理

In [1]:
import os, time, json, warnings

# リポジトリルートを作業ディレクトリにして、data/ などの相対パスを揃える
if os.path.basename(os.getcwd()) == "notebooks":
    os.chdir("..")

import numpy as np
import pandas as pd
from sklearn.metrics import roc_auc_score
from sklearn.model_selection import StratifiedKFold
from sklearn.preprocessing import KBinsDiscretizer, TargetEncoder

pd.set_option("display.width", 160)
pd.set_option("display.max_columns", 20)

# 絶対パスは出力しない設計(公開リポジトリに対応)
print("data/ を検出:", os.path.isdir("data"))

data/ を検出: True


In [2]:
train = pd.read_csv("data/train.csv")
test = pd.read_csv("data/test.csv")
TARGET = "Will_Buy_EV"
y = (train[TARGET] == "Yes").astype(int).to_numpy()

NUM = ["Age", "Annual_Income_USD", "Daily_Commute_km", "Number_of_Cars_Owned",
       "Charging_Stations_Near_Home", "Charging_Stations_Near_Work", "Environmental_Concern_Level"]
CAT = ["Gender", "City_Type", "Current_Car_Type", "Home_Charging_Possible",
       "Subsidy_Available", "Range_Anxiety_Level"]

# 値の種類（ユニーク値）が少ない数値列(CatBoost の catify で使う。4-3)
LOWCARD_NUM = ["Age", "Number_of_Cars_Owned", "Charging_Stations_Near_Home",
               "Charging_Stations_Near_Work", "Environmental_Concern_Level"]

# fold 分割は全モデル共通(変更禁止)。Feature Engineering の実演には fold 1 を使う
skf = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)
tr_idx, va_idx = next(iter(skf.split(train, y)))

print(f"train {train.shape} / test {test.shape} / 購入率 {y.mean():.3f}")
print(f"欠損値: train {int(train.isna().sum().sum())} 個 / test {int(test.isna().sum().sum())} 個(補完は不要)")
print(f"fold 1: 学習 {len(tr_idx):,} 行 / 検証 {len(va_idx):,} 行")
print("数値列    :", NUM)
print("カテゴリ列:", CAT)

train (668665, 15) / test (286571, 14) / 購入率 0.175
欠損値: train 0 個 / test 0 個(補完は不要)
fold 1: 学習 534,932 行 / 検証 133,733 行
数値列    : ['Age', 'Annual_Income_USD', 'Daily_Commute_km', 'Number_of_Cars_Owned', 'Charging_Stations_Near_Home', 'Charging_Stations_Near_Work', 'Environmental_Concern_Level']
カテゴリ列: ['Gender', 'City_Type', 'Current_Car_Type', 'Home_Charging_Possible', 'Subsidy_Available', 'Range_Anxiety_Level']


各章の最後で、組み上げた列が本番と同じかを確かめる。

本番の列名は `04_train_and_evaluate_<model>.py --dump-features` が書き出した `docs/features_<model>.json` にある。

In [3]:
def check_columns(tag, columns):
    # 作った列が本番(04 の --dump-features が書き出した docs/features_<tag>.json)と一致するかを確かめる
    with open(f"docs/features_{tag}.json", encoding="utf-8") as f:
        expected = json.load(f)["columns"]
    cols = list(columns)
    same_set = set(cols) == set(expected)
    same_order = cols == expected

    print(f"{tag}: {len(cols)} 列(本番 {len(expected)} 列)"
          f" → 列の集合 {'一致' if same_set else '不一致'} / 並び順 {'一致' if same_order else '不一致'}")
    if not same_set:
        print("  ノートブックだけにある列:", sorted(set(cols) - set(expected)))
        print("  本番だけにある列      :", sorted(set(expected) - set(cols)))

---

## 3. GBDT 3種(LightGBM・XGBoost・CatBoost)で共通の Feature Engineering

3種の木モデルに共通する Feature Engineering。**RealMLP は前処理が別系統なので 5 章で扱う。**

この章では LightGBM 版で作る。XGBoost・CatBoost の版は考え方が同じで実装の細部だけが違うので、4 章で差分を示して作る。

### 3-1 厳密値キー — Target Encoding・Count の材料になるキーを作る

> 仮説2の検証

**ここではキーを作るだけで、購入率はまだ計算しない**(目的変数は使わない)。

このキーを使って、購入率を計算する Target Encoding を 3-4 で、出現回数を数える Count を 3-6 で作る。

- **数値列も丸めずに、値のままキーにする。** キーを作る時点で丸めると、値ごとの違いが消える(EDA の仮説2)
- 数値列は読み込んだ値をそのままキーにする(同じ CSV から読んだ同じ値は、浮動小数でも必ず同じ値になる)
- カテゴリ列は整数コードにする。コードの割り振り(水準の一覧)は train と test を合わせて決める

下では、キーごとに「値の種類数」と「1値あたりの行数」を数える。行数が少ない値ほど、3-4 で計算する購入率が不安定になる。

In [4]:
# カテゴリ列の水準は train と test を合わせて決める(目的変数を使わないのでリークしない)
cat_levels = {c: pd.concat([train[c], test[c]]).astype("category").cat.categories for c in CAT}


def make_key_frame(df):
    # 13 列を TE / Count のキーにする。数値列は値そのまま、カテゴリ列は整数コード
    keys = pd.DataFrame(index=df.index)
    for c in NUM:
        keys[c] = df[c].to_numpy()
    for c in CAT:
        keys[c] = pd.Categorical(df[c], categories=cat_levels[c]).codes.astype("int16")
    return keys


keys_tr, keys_te = make_key_frame(train), make_key_frame(test)

rows_per_value = {c: keys_tr[c].value_counts() for c in keys_tr.columns}
summary = pd.DataFrame({
    "種類数": keys_tr.nunique(),
    "1値あたりの行数(平均)": {c: round(v.mean(), 1) for c, v in rows_per_value.items()},
    "1値あたりの行数(中央値)": {c: v.median() for c, v in rows_per_value.items()},
})

print(summary.to_string())

inc_counts = rows_per_value["Annual_Income_USD"]

print()
print(f"年収: 10 行未満の値が {(inc_counts < 10).mean():.1%} / "
      f"最小値 {train['Annual_Income_USD'].min():,.0f} ドルだけで全体の {(train['Annual_Income_USD'] == train['Annual_Income_USD'].min()).mean():.1%}")

                               種類数  1値あたりの行数(平均)  1値あたりの行数(中央値)
Age                             45       14859.2        14600.0
Annual_Income_USD            13214          50.6            8.0
Daily_Commute_km               805         830.6          576.0
Number_of_Cars_Owned             4      167166.2       178444.5
Charging_Stations_Near_Home     15       44577.7        52986.0
Charging_Stations_Near_Work     20       33433.2        22861.0
Environmental_Concern_Level      5      133733.0       130469.0
Gender                           3      222888.3       295427.0
City_Type                        3      222888.3       255377.0
Current_Car_Type                 4      167166.2       162991.5
Home_Charging_Possible           2      334332.5       334332.5
Subsidy_Available                2      334332.5       334332.5
Range_Anxiety_Level              3      222888.3        62499.0

年収: 10 行未満の値が 50.8% / 最小値 30,000 ドルだけで全体の 9.2%


**年収は「1値あたり平均 50 行」だが、中央値は 8 行しかない。** 平均を押し上げているのは、

最小値 30,000 ドルに全体の 9% が集まっている部分。

**半分の値は 10 行未満**で、その値だけの購入率はノイズが大きい。これを補うのが次の Smooth Keys。

### 3-2 Smooth Keys — 同じ年収を、粗い解像度でも見せる

**意図**: 厳密値キーだけだと、行数の少ない値(年収の半分)は購入率の推定が不安定になる。

そこで「年収 /100(100 ドル刻み)」「/1000」のように**近い値をまとめたキーも併せて**持たせ、行数の少ない値でも近い値の購入率を参照できるようにする。

**厳密値キーを置き換えるのではなく足す。** 

改善幅は単独では測っておらず、平滑化3種の Target Encoding・digit・ビン数 1024 とまとめて Run 9 で採用した(+0.00048〜0.00096)。

| モデル | 年収 | 通勤距離 | 使い方 |
|---|---|---|---|
| LightGBM | /10・/100・/1000 | floor(km) | Target Encoding / Count のキー |
| XGBoost | /100・/1000・/10000 | floor(km) | Target Encoding のキー |
| CatBoost | /100・/1000 | floor(km) | Target Encoding のキー |

下は LightGBM 版。

In [5]:
def add_smooth_keys(keys, df):
    # 年収を 10 / 100 / 1000 ドル刻み、通勤距離を 1 km 刻みに丸めたキーを足す(LightGBM 版)
    keys = keys.copy()
    inc = df["Annual_Income_USD"].to_numpy(dtype="float64")
    km = df["Daily_Commute_km"].to_numpy(dtype="float64")
    keys["sk_inc10"] = np.floor(inc / 10.0).astype("int32")
    keys["sk_inc100"] = np.floor(inc / 100.0).astype("int32")
    keys["sk_inc1000"] = np.floor(inc / 1000.0).astype("int32")
    keys["sk_commute"] = np.floor(km).astype("int32")
    return keys


SMOOTH_KEYS = ["sk_inc10", "sk_inc100", "sk_inc1000", "sk_commute"]
keys_tr, keys_te = add_smooth_keys(keys_tr, train), add_smooth_keys(keys_te, test)

sample = pd.concat([train[["Annual_Income_USD", "Daily_Commute_km"]], keys_tr[SMOOTH_KEYS]], axis=1)

print(sample.head().to_string())
print()
print("種類数:", keys_tr[SMOOTH_KEYS].nunique().to_dict())
print("1値あたりの行数(中央値):", {c: int(keys_tr[c].value_counts().median()) for c in SMOOTH_KEYS})

   Annual_Income_USD  Daily_Commute_km  sk_inc10  sk_inc100  sk_inc1000  sk_commute
0            92887.0              23.4      9288        928          92          23
1            30000.0               5.0      3000        300          30           5
2            94389.0              36.8      9438        943          94          36
3            73580.0              23.7      7358        735          73          23
4            57898.0              50.8      5789        578          57          50

種類数: {'sk_inc10': 5933, 'sk_inc100': 1142, 'sk_inc1000': 150, 'sk_commute': 89}
1値あたりの行数(中央値): {'sk_inc10': 55, 'sk_inc100': 298, 'sk_inc1000': 3088, 'sk_commute': 5429}


年収の1値あたりの行数(中央値)は、厳密値の 8 行から /10 で 55 行、/100 で約 300 行、/1000 で約 3,000 行に増える。

粗くするほど推定は安定するが、値ごとの違いは見えにくくなる。その両方を並べて渡している。

### 3-3 どの列に、どのエンコーディングを作るか

厳密値キー(3-1)と Smooth Keys(3-2)をもとに、3種類のエンコーディング(digit・Target Encoding・Count)を作る。

**ただし全部の列に全部を作るのではなく、情報を足すものだけを作る。** LightGBM 版の設計は次のとおり。

| 列のグループ | digit(3-5) | Target Encoding(3-4) | Count(3-6) |
|---|---|---|---|
| 年収・通勤距離(値の種類（ユニーク値）(専門用語ではカーディナリティ)が多い数値列) | ○ | ○ 平滑化3種 | ○ |
| Smooth Keys 4本 | — | ○ 平滑化3種 | — |
| 値の種類（ユニーク値）が少ない数値列(年齢・保有台数・スタンド数2列・環境意識) | — | ○ 平滑化1種(auto) | — |
| カテゴリ列6列 | — | — | — |
| 組み合わせ: 自宅充電の可否 × 自宅スタンド数 | — | ○ 平滑化1種(auto) | — |

- **平滑化3種が意味を持つのは、行数の少ない値を持つキーだけ。** 値の種類（ユニーク値）が少ない列は1値あたり数万行あり、3種が同じ値になる
- **値の種類（ユニーク値）が少ない数値列は Target Encoding だけで足りる。** Target Encoding が値ごとの購入率を1列で渡すので、並び順を変えるだけの digit・Count は情報を足さない。Target Encoding を作らないと有意に悪化する
- **カテゴリ列は、LightGBM がカテゴリのまま好きな組に分けられる**ので、Target Encoding も Count も要らない
- **組み合わせは、EDA で見つかった唯一の交互作用の1組だけ**。木は自力で表現できるので効果は誤差(+0.000012)だが、CV 最高の構成として採用
- XGBoost・CatBoost はカテゴリの扱いが違うので、作る列も少し違う(4-2・4-3)

In [6]:
# どの列に、どのエンコーディングを作るか(LightGBM 版)
HIGH_CARD = ["Annual_Income_USD", "Daily_Commute_km"]   # 値の種類（ユニーク値）が多い数値列
TRIPLE = ("auto", 10.0, 100.0)                          # 平滑化3種

DIGIT_COLS = HIGH_CARD                                  # digit を作る列(3-5)
COUNT_COLS = HIGH_CARD                                  # Count を作る列(3-6)

# Target Encoding のキーと平滑化(3-4)。並びは生の数値列の順 → Smooth Keys
# 組み合わせキー: 自宅充電の可否 × 自宅スタンド数(2 列の値を文字列でつなぐ)
PAIR = "Home_Charging_PossibleXCharging_Stations_Near_Home"

for keys in (keys_tr, keys_te):
    keys[PAIR] = keys["Home_Charging_Possible"].astype(str) + "_" + keys["Charging_Stations_Near_Home"].astype(str)

TE_PLAN = ([(c, TRIPLE if c in HIGH_CARD else ("auto",)) for c in NUM]
           + [(k, TRIPLE) for k in SMOOTH_KEYS]
           + [(PAIR, ("auto",))])

plan = pd.DataFrame([(k, " / ".join(map(str, s))) for k, s in TE_PLAN], columns=["TE のキー", "平滑化"])

print(plan.to_string(index=False))
print(f"\n作る列: digit は {DIGIT_COLS} / Count は {COUNT_COLS} / TE は {sum(len(s) for _, s in TE_PLAN)} 列")

                                            TE のキー                 平滑化
                                               Age                auto
                                 Annual_Income_USD auto / 10.0 / 100.0
                                  Daily_Commute_km auto / 10.0 / 100.0
                              Number_of_Cars_Owned                auto
                       Charging_Stations_Near_Home                auto
                       Charging_Stations_Near_Work                auto
                       Environmental_Concern_Level                auto
                                          sk_inc10 auto / 10.0 / 100.0
                                         sk_inc100 auto / 10.0 / 100.0
                                        sk_inc1000 auto / 10.0 / 100.0
                                        sk_commute auto / 10.0 / 100.0
Home_Charging_PossibleXCharging_Stations_Near_Home                auto

作る列: digit は ['Annual_Income_USD', 'Daily_Commute_km'] / Count は ['Annual_In

### 3-4 厳密値 Target Encoding(GBDT 3種)— 最大の改善要因(仮説2)

**意図**: 3-1 のキーごとに「**その値の人の購入率**」を計算し、1列にして渡す。

年収をそのまま渡すと、木は値ごとの違いを学べない。木は学習前に値を 255 個のビンにまとめるので、年収 13,214 種類では平均 52 種類の値が同じ扱いになる。また分割は「年収 ≧ ○○ か」という大小の境目だけなので、値ごとに上下する購入率を切り分けるには境目が何本も要る。

購入率の列なら、**「購入率 ≧ 20% か」の1回の分割で、購入率の高い値と低い値をまとめて分けられる。**

改善幅は **+0.00345**(LightGBM、3-fold)。前コンペ(S6E8)でも同じ施策が唯一桁違いに効いた。

**目的変数を使うので、リークを防ぐ作法が要る。** それぞれの理由も書いておく。

| 作法 | やり方 | なぜ必要か |
|---|---|---|
| fold 内で作る | fold 1 の学習行だけで購入率を集計し、検証行・test にはその値を当てる | 検証行の正解が混ざると、検証スコアが実際より良く見える |
| **自分の正解を使わずに計算する**(一般に **Out-of-Fold** と呼ばれる) | 学習行自身の値は、さらに内側で 5 分割して**自分を含まない行から**計算する | 自分の正解を含む購入率は、学習時だけ当たりすぎる。木がそれを過信する。Out-of-Fold にすると **+0.00108**(XGBoost) |
| 平滑化 | 行数の少ない値は、全体の購入率(17.5%)に寄せる | 3 行しかない値の購入率 0% や 100% を、そのまま信じないため |
| **平滑化を3種類同時に入れる** | 強さ auto / 10 / 100 の3列を並べる(選ぶのではない) | 行数の多い値では弱い平滑化、少ない値では強い平滑化を、**木が値ごとに使い分けられる**。ただし意味があるのは、年収・通勤距離のように**行数の少ない値を持つキーだけ**(3-3) |

LightGBM で作るのは、年収・通勤距離と Smooth Keys 4本の 6 キー × 平滑化3種 = 18 列と、値の種類（ユニーク値）が少ない数値列 5 キー × auto = 5 列、組み合わせ(自宅充電 × 自宅スタンド数)の auto 1 列の **24 列**(3-3 の TE_PLAN)。

> **注意: 木を弱くしてからでないと逆効果になる。** 深さ無制限・全列を使う設定では、
> 同じ Target Encoding が **-0.00831** と大きく悪化した。年収の Target Encoding はノイズを含むので、強い木はそれを目印に葉を切り刻む。
> 本番の「深さ 5 + 1本の木が見る列を 30% に絞る」で、1本の木が Target Encoding に依存しきれない状態にして初めて効く。

下は LightGBM 版(3モデルとも考え方は同じ)。

In [7]:
def _agg(keys, y_part):
    # キーごとの (購入者数, 行数)。平滑化の強さによらない部分なので、3 種類で使い回す
    return pd.DataFrame({"k": keys, "y": y_part}).groupby("k", observed=True)["y"].agg(["sum", "count"])


def _smoothed_rate(agg, prior, smooth):
    # 平滑化した購入率 = (購入者数 + 全体の購入率 × m) / (行数 + m)
    # smooth="auto" は m を値ごとに決める(sklearn の TargetEncoder と同じ経験ベイズの規則)
    cnt = agg["count"].to_numpy(dtype="float64")
    s = agg["sum"].to_numpy(dtype="float64")
    if smooth == "auto":
        p_i = s / cnt
        m = (p_i * (1.0 - p_i)) / (prior * (1.0 - prior))
    else:
        m = float(smooth)
    return pd.Series((s + prior * m) / (cnt + m), index=agg.index)


def target_encode(keys_fit, y_fit, others, plan, n_inner=5, seed=42):
    # Out-of-Fold の Target Encoding(LightGBM 版)
    #   学習行      : 内側でさらに 5 分割し、自分を含まない 4/5 の行で計算した値
    #   検証行・test: 学習行すべてで計算した値
    # plan: [(キー, 平滑化の種類), ...]。キーごとに作る平滑化を変えられる
    prior = float(y_fit.mean())
    te_fit = pd.DataFrame(index=keys_fit.index)
    te_others = [pd.DataFrame(index=f.index) for f in others]
    inner = list(StratifiedKFold(n_splits=n_inner, shuffle=True, random_state=seed)
                 .split(np.zeros(len(y_fit)), y_fit))
    for c, smooths in plan:
        arr = keys_fit[c].to_numpy()
        names = {sm: f"te_{c}_s{sm if sm == 'auto' else int(sm)}" for sm in smooths}
        vals = {sm: np.full(len(arr), prior, dtype="float32") for sm in smooths}
        for in_idx, out_idx in inner:
            agg = _agg(arr[in_idx], y_fit[in_idx])
            s_out = pd.Series(arr[out_idx])
            for sm in smooths:
                rate = _smoothed_rate(agg, prior, sm)
                vals[sm][out_idx] = s_out.map(rate).fillna(prior).to_numpy(dtype="float32")
        for sm in smooths:
            te_fit[names[sm]] = vals[sm]
        agg_full = _agg(arr, y_fit)
        for sm in smooths:
            rate = _smoothed_rate(agg_full, prior, sm)
            for f, out in zip(others, te_others):
                out[names[sm]] = f[c].map(rate).fillna(prior).to_numpy(dtype="float32")
    return te_fit, te_others


t0 = time.time()
te_tr, (te_va, te_te) = target_encode(keys_tr.iloc[tr_idx], y[tr_idx],
                                      [keys_tr.iloc[va_idx], keys_te], TE_PLAN)

print(f"{len(TE_PLAN)} キー → TE {te_tr.shape[1]} 列({time.time() - t0:.0f} 秒)")
print()
print(te_va.iloc[:3, :6].round(4).to_string())

12 キー → TE 24 列(3 秒)

   te_Age_sauto  te_Annual_Income_USD_sauto  te_Annual_Income_USD_s10  te_Annual_Income_USD_s100  te_Daily_Commute_km_sauto  te_Daily_Commute_km_s10
1        0.1514                      0.0445                    0.0445                     0.0447                     0.1842                   0.1842
2        0.2068                      0.4005                    0.3622                     0.2408                     0.1902                   0.1901
6        0.2064                      0.1746                    0.1746                     0.1746                     0.1816                   0.1815


**リークがないことの確認。** 学習行に自分の正解が混ざっていれば、学習行での AUC だけが不自然に高くなる。

学習行と検証行で、Target Encoding 列1本だけの AUC がほぼ同じなら問題ない。

In [8]:
auto_cols = [c for c in te_va.columns if "auto" in c]
check = pd.DataFrame({
    "学習行での AUC": [roc_auc_score(y[tr_idx], te_tr[c]) for c in auto_cols],
    "検証行での AUC": [roc_auc_score(y[va_idx], te_va[c]) for c in auto_cols],
}, index=auto_cols).sort_values("検証行での AUC", ascending=False)

print(check.round(4).head(8).to_string())

                                                             学習行での AUC  検証行での AUC
te_Environmental_Concern_Level_sauto                            0.8435     0.8409
te_sk_inc10_sauto                                               0.7096     0.7137
te_Annual_Income_USD_sauto                                      0.7070     0.7120
te_sk_inc100_sauto                                              0.7031     0.7039
te_sk_inc1000_sauto                                             0.6841     0.6830
te_Home_Charging_PossibleXCharging_Stations_Near_Home_sauto     0.5626     0.5649
te_sk_commute_sauto                                             0.5490     0.5533
te_Daily_Commute_km_sauto                                       0.5488     0.5530


学習行と検証行の AUC の差は最大でも 0.005 程度で、学習行だけが高いという偏りもない。Out-of-Fold が効いている。

### 3-5 digit features — 数値を桁ごとの列にばらす

**意図**: 木モデルは学習前に数値列を**ビン**(LightGBM・XGBoost は既定 255 個、CatBoost は本番の軽量設定で 64 個)にまとめ、

ビンの境目でしか分割しない。年収 13,214 種類は 255 個に押し込まれ、**平均 52 種類の値が同じ扱い**になる。

年収 92,887 を「1の位 7 / 10の位 8 / 100の位 8 / 1000の位 2 …」と桁ごとの列にすると、

**各列は 0〜9 の10種類しかないのでビンで潰れない。** 木は桁の列を順に分割していけば、元の値まで辿り着ける。

| モデル | 列数 | 違い |
|---|---|---|
| LightGBM | 7(年収 4・通勤距離 3。値の種類（ユニーク値）が少ない列には作らない。3-3) | 浮動小数で計算し、train で定数の桁を落とす |
| XGBoost / CatBoost | XGBoost 14 / CatBoost 10(4-2・4-3) | 整数に直してから計算し、train と test の両方で定数の桁だけ落とす |

**正直な測定結果(Run 14、平滑化3種の Target Encoding なしで digit だけを足した場合)**

| モデル | digit のみ | ビン数を上げるだけ | 両方 |
|---|---|---|---|
| LightGBM(ビン 255 → 1024) | +0.00010 | +0.00001 | +0.00011 |
| XGBoost(ビン 255 → 1024) | +0.00009 | ±0 | +0.00007 |
| CatBoost(ビン 64 → 1024) | **+0.00044** | **+0.00038** | **+0.00061** |

**LightGBM・XGBoost では単独の効果は誤差だった。** 厳密値 Target Encoding が値ごとの購入率を既に渡しているため。

**CatBoost でははっきり効いた。** ビンが 64 個と最も粗く、年収の値の違いが最も潰れていたため。

LightGBM・XGBoost でも入れたままにしているのは、悪化はせず、Run 9 の一式(+0.0005)の一部として採用したから。

下は LightGBM 版。

In [9]:
DIGIT_K = list(range(-4, 4))   # 10^-4 の位 〜 10^3 の位


def add_digit_features(df, cols, keep=None):
    # (x // 10^k) % 10 を列にする(LightGBM 版: 浮動小数で計算し、train で定数になる桁は落とす)
    out = {}
    for c in cols:
        x = df[c].to_numpy(dtype="float64")
        for k in DIGIT_K:
            name = f"{c}_digit{k}"
            if keep is not None and name not in keep:
                continue
            arr = np.mod(np.floor(x * (10.0 ** (-k)) + 1e-9), 10.0).astype("int8")
            if keep is None and arr.min() == arr.max():   # 定数の桁は情報がないので落とす
                continue
            out[name] = arr
    return pd.DataFrame(out, index=df.index)


digit_tr = add_digit_features(train, DIGIT_COLS)
digit_te = add_digit_features(test, DIGIT_COLS, keep=list(digit_tr.columns))   # test も train と同じ列にそろえる

inc_digits = [c for c in digit_tr.columns if c.startswith("Annual_Income_USD")]

print(pd.concat([train["Annual_Income_USD"], digit_tr[inc_digits]], axis=1).head().to_string())
print()
print(f"digit 列: {digit_tr.shape[1]} 列 →", list(digit_tr.columns))

   Annual_Income_USD  Annual_Income_USD_digit0  Annual_Income_USD_digit1  Annual_Income_USD_digit2  Annual_Income_USD_digit3
0            92887.0                         7                         8                         8                         2
1            30000.0                         0                         0                         0                         0
2            94389.0                         9                         8                         3                         4
3            73580.0                         0                         8                         5                         3
4            57898.0                         8                         9                         8                         7

digit 列: 7 列 → ['Annual_Income_USD_digit0', 'Annual_Income_USD_digit1', 'Annual_Income_USD_digit2', 'Annual_Income_USD_digit3', 'Daily_Commute_km_digit-1', 'Daily_Commute_km_digit0', 'Daily_Commute_km_digit1']


### 3-6 Count Encoding(LightGBM・XGBoost)— +0.00083 / +0.00049

**意図**: 各キーの値が**何回出てくるか**(頻度)を列にする。役割は2つある。

- **Target Encoding の信頼度を木に教える。** 同じ購入率 20% でも、1,000 行から出た 20% と 3 行から出た 20% では重みが違う。
  頻度の列があれば、木は「行数が多い値の Target Encoding だけ信じる」という分割ができる
- **頻度そのものに生成過程の痕跡がある。** 年収 30,000 ドルは全体の 9% を占める(3-1)。
  頻度の列はこうした特別な値を1回の分割で切り出せる

目的変数を使わないので、train と test をまとめて数えてよい(リークしない)。

Target Encoding と併用すると効果が足し算になった(Target Encoding のみ +0.00345 → 併用 +0.00395)。**Target Encoding とは別の情報を運んでいる**裏付け。

CatBoost では **-0.00017** と効かなかった。CatBoost は内部で同種の統計を自動で作るため。

LightGBM で作るのは、年収・通勤距離の2列だけ(3-3)。

In [10]:
def count_encode(keys_tr, keys_te, cols):
    # 値の出現頻度(train と test を合わせて数え、全行数で割る)。目的変数を使わないのでリークしない
    out_tr, out_te = pd.DataFrame(index=keys_tr.index), pd.DataFrame(index=keys_te.index)
    n_total = len(keys_tr) + len(keys_te)
    for c in cols:
        vc = pd.concat([keys_tr[c], keys_te[c]], ignore_index=True).value_counts()
        out_tr[f"cnt_{c}"] = (keys_tr[c].map(vc).astype("float32") / n_total).to_numpy()
        out_te[f"cnt_{c}"] = (keys_te[c].map(vc).astype("float32") / n_total).to_numpy()
    return out_tr, out_te


cnt_tr, cnt_te = count_encode(keys_tr, keys_te, COUNT_COLS)

print(pd.concat([train[COUNT_COLS], cnt_tr], axis=1).head().to_string())
print()
print(f"Count 列: {cnt_tr.shape[1]} 列")

   Annual_Income_USD  Daily_Commute_km  cnt_Annual_Income_USD  cnt_Daily_Commute_km
0            92887.0              23.4               0.000156              0.001025
1            30000.0               5.0               0.091999              0.215599
2            94389.0              36.8               0.000063              0.002892
3            73580.0              23.7               0.000263              0.000950
4            57898.0              50.8               0.000433              0.001025

Count 列: 2 列


### 3-7(補足)厳密値・Smooth Keys・digit は矛盾しないのか

「値を丸めずに使う」と「粗く丸めたキー」や「桁への分解」という処理は矛盾に見えるが、**3つは同じ目的(値の細部を失わせない)に対する、弱点の違う経路**として併用している。

| 部品 | 何を渡すか | 強み | 弱み(次の部品が補う) |
|---|---|---|---|
| 厳密値キー → Target Encoding | 「その値ちょうど」の購入率 | 値ごとの違いを1列で直接渡せる | 行数の少ない値(年収の半分は 10 行未満)では推定がぶれる |
| Smooth Keys → Target Encoding | 「近い値の範囲」の購入率 | 行数が増えて推定が安定する | 値ごとの細かい違いは見えない |
| digit features | 値そのもの(ラベルを使わない形) | ビンで潰れない。**目的変数を通さない**ので Target Encoding のぶれと無関係 | 1列だけでは意味を持たず、効果も小さい(CatBoost 以外は誤差) |

3つとも並べて渡し、**値ごとにどれを信じるかは木が選ぶ**(平滑化を3種類並べる 3-4 も同じ考え方)。

---

## 4. GBDT のモデル別 Feature Engineering

3 章の部品に、モデル固有の処理を足して fold 1 の学習行列を組み上げ、本番の列と突き合わせる。

### 4-1 LightGBM — カテゴリ列を native category で渡す

**意図**: カテゴリ列(2〜4種類)を `category` 型のまま渡すと、LightGBM は値をグループに分けて分割できる。

整数コード(ordinal)でも測ったが **-0.000005(z=-0.42)で差がなかった**(Run 21)。

カテゴリ列の情報は Target Encoding 経由でも渡っているため、渡し方の違いが効かない。最も素直な native category のままにした。

組み上げ = 生の13列 + digit 7 列(3-5)+ Count 2 列(3-6)+ Target Encoding 24 列(3-4)= **46 列**。

In [11]:
# カテゴリ列を category 型にする(水準は 3-1 と同じく train と test を合わせたもの)
train_c, test_c = train.copy(), test.copy()

for c in CAT:
    train_c[c] = pd.Categorical(train[c], categories=cat_levels[c])
    test_c[c] = pd.Categorical(test[c], categories=cat_levels[c])

X_static = pd.concat([train_c[NUM + CAT], digit_tr, cnt_tr], axis=1)
X_static_te = pd.concat([test_c[NUM + CAT], digit_te, cnt_te], axis=1)

X_tr_lgbm = pd.concat([X_static.iloc[tr_idx].reset_index(drop=True), te_tr.reset_index(drop=True)], axis=1)
X_va_lgbm = pd.concat([X_static.iloc[va_idx].reset_index(drop=True), te_va.reset_index(drop=True)], axis=1)
X_te_lgbm = pd.concat([X_static_te.reset_index(drop=True), te_te.reset_index(drop=True)], axis=1)

print("category 型で渡す列:", [c for c in X_tr_lgbm.columns if str(X_tr_lgbm[c].dtype) == "category"])
check_columns("lgbm", X_tr_lgbm.columns)

category 型で渡す列: ['Gender', 'City_Type', 'Current_Car_Type', 'Home_Charging_Possible', 'Subsidy_Available', 'Range_Anxiety_Level']
lgbm: 46 列(本番 46 列) → 列の集合 一致 / 並び順 一致


### 4-2 XGBoost — カテゴリ列を整数コード(ordinal)で渡す

**意図: 精度ではなく多様性のため。** native category / ordinal / One-Hot の3方式は単体スコアが同点だった

(ordinal +0.00009、One-Hot +0.00005 で、どれも誤差)。

アンサンブルでは**同じ予測をするモデルを2つ入れても伸びない**ので、LightGBM(native category)と

違う渡し方にして、予測の癖をずらした。

**部品の考え方は 3 章と同じだが、実装の細部が LightGBM 版と違う。**

| 部品 | LightGBM 版(3 章) | XGBoost 版 |
|---|---|---|
| Smooth Keys | 年収 /10・/100・/1000 | 年収 **/100・/1000・/10000** |
| digit | 浮動小数で計算。年収・通勤距離だけ(7 列) | **1万倍して整数にしてから計算**。値の種類（ユニーク値）が少ない列にも作る(14 列)。保有台数・環境意識は値が 0〜9 なので1の位が値そのものになり、作らない |
| Target Encoding の基準値 | 内側の分割でも、学習行全体の購入率に寄せる | 内側の分割では、**その分割の学習部分の購入率**に寄せる |
| Target Encoding の列名 | `te_<キー>_sauto` など | `<キー>_tea` など |
| Count | 年収・通勤距離の2列。列名 `cnt_<列>` | **全13列**(カテゴリを整数コードで渡すので、Count が並び順を変える役に立つ)。列名 `<列>_ce`、値を文字列にして数える |
| Target Encoding | 値の種類（ユニーク値）が少ない数値列は auto だけ、カテゴリ列には作らない | 値の種類（ユニーク値）が少ない11キー(カテゴリ列を含む)は auto だけ、他は平滑化3種 |

> XGBoost でも LightGBM と同じように絞ると、単体 -0.00003、アンサンブル -0.00001(z=-2.5)と悪化傾向だったので、XGBoost は値の種類（ユニーク値）が少ない列の digit・Count・カテゴリ列の Target Encoding も作る。

In [12]:
ALL = NUM + CAT
DIGIT_KS = range(-4, 4)


def digit_block(df, cols):
    # 桁の列(XGBoost・CatBoost 版): 1 万倍して整数にしてから計算するので、浮動小数の誤差が桁に漏れない
    out = {}
    for c in cols:
        iv = np.rint(df[c].astype("float64").to_numpy() * 10_000).astype("int64")
        for k in DIGIT_KS:
            out[f"{c}_d{k}"] = (iv // (10 ** (k + 4))) % 10
    return pd.DataFrame(out, index=df.index).astype("int8")


# ① 生の 13 列 + digit(保有台数・環境意識は 1 の位が値そのものなので作らない。定数になる桁は落とす)
XGB_DIGIT_COLS = [c for c in NUM if c not in ("Number_of_Cars_Owned", "Environmental_Concern_Level")]
d_tr, d_te = digit_block(train, XGB_DIGIT_COLS), digit_block(test, XGB_DIGIT_COLS)
digit_keep = [c for c in d_tr.columns if d_tr[c].nunique() > 1 or d_te[c].nunique() > 1]
xtr = pd.concat([train[ALL].reset_index(drop=True), d_tr[digit_keep].reset_index(drop=True)], axis=1)
xte = pd.concat([test[ALL].reset_index(drop=True), d_te[digit_keep].reset_index(drop=True)], axis=1)

# ② Count(値を文字列にして train + test で数え、全行数で割る)
n_total = len(train) + len(test)

for c in ALL:
    freq = pd.concat([train[c].astype(str), test[c].astype(str)], ignore_index=True).value_counts() / n_total
    xtr[f"{c}_ce"] = train[c].astype(str).map(freq).astype("float32")
    xte[f"{c}_ce"] = test[c].astype(str).map(freq).astype("float32")

# ③ カテゴリ列を整数コード(ordinal)に
for c in CAT:
    levels = pd.concat([xtr[c].astype(str), xte[c].astype(str)]).astype("category").cat.categories
    mapping = {v: i for i, v in enumerate(levels)}
    xtr[c] = xtr[c].astype(str).map(mapping).astype("int16")
    xte[c] = xte[c].astype(str).map(mapping).astype("int16")

print("カテゴリ列の中身(整数コード):", {c: sorted(xtr[c].unique().tolist()) for c in ["City_Type", "Range_Anxiety_Level"]})
print(f"digit {len(digit_keep)} 列 / Count {len(ALL)} 列")

カテゴリ列の中身(整数コード): {'City_Type': [0, 1, 2], 'Range_Anxiety_Level': [0, 1, 2]}
digit 14 列 / Count 13 列


In [13]:
# ④ Target Encoding のキー: Smooth Keys 4 本(XGBoost の刻み)+ 生の 13 列。値の種類（ユニーク値）が少ない 11 キーは auto だけ
XGB_SMOOTH = [("inc_f100", "Annual_Income_USD", 100), ("inc_f1000", "Annual_Income_USD", 1000),
              ("inc_f10000", "Annual_Income_USD", 10000), ("commute_f1", "Daily_Commute_km", 1)]


def smooth_key_frame(df, specs):
    # floor(値 / 刻み) のキーを作る
    return pd.DataFrame({name: np.floor(df[col].astype("float64").to_numpy() / div).astype("int64")
                         for name, col, div in specs}, index=df.index)


src_tr = pd.concat([smooth_key_frame(train, XGB_SMOOTH), train[ALL]], axis=1)
src_te = pd.concat([smooth_key_frame(test, XGB_SMOOTH), test[ALL]], axis=1)
xgb_te_cols = [name for name, _, _ in XGB_SMOOTH] + ALL
XGB_TE_PLAN = ([(k, TRIPLE) for k in xgb_te_cols[:4]]
               + [(c, TRIPLE if c in HIGH_CARD else ("auto",)) for c in ALL])

# キーを train + test で整数コードにそろえておく(集計を numpy の bincount で速く回すため)
codes_tr, codes_te, ncats = {}, {}, {}

for c in xgb_te_cols:
    codes, uniques = pd.factorize(pd.concat([src_tr[c], src_te[c]], ignore_index=True), sort=False)
    codes = codes.astype("int32")
    codes_tr[c], codes_te[c], ncats[c] = codes[:len(train)], codes[len(train):], len(uniques)


def _encode_codes(codes, y_part, ncat, smooths, prior, y_var):
    # 整数コードごとの平滑化した購入率(XGBoost 版)。"auto" は LightGBM 版と同じ規則を別の形で書いたもの
    cnt = np.bincount(codes, minlength=ncat).astype("float64")
    s = np.bincount(codes, weights=y_part, minlength=ncat)
    mean = s / np.maximum(cnt, 1.0)
    out = []
    for sm in smooths:
        if sm == "auto":
            var_cat = mean * (1.0 - mean)
            denom = y_var * cnt + var_cat
            lam = np.where(denom > 0, (y_var * cnt) / np.where(denom > 0, denom, 1.0), 0.0)
            enc = lam * mean + (1.0 - lam) * prior
        else:
            enc = (s + prior * float(sm)) / (cnt + float(sm))
        out.append(np.where(cnt > 0, enc, prior).astype("float32"))
    return out


def target_encode_xgb(plan, n_inner=5, seed=42):
    # Out-of-Fold の Target Encoding(XGBoost 版)。内側の分割では、その分割の購入率を基準値にする
    # plan: [(キー, 平滑化の種類), ...]
    fit_y = y[tr_idx].astype("float64")
    names = [f"{c}_te{'a' if sm == 'auto' else int(sm)}" for c, smooths in plan for sm in smooths]
    te_fit = np.zeros((len(tr_idx), len(names)), dtype="float32")
    inner = StratifiedKFold(n_splits=n_inner, shuffle=True, random_state=seed)
    for in_tr, in_va in inner.split(np.zeros((len(tr_idx), 1)), fit_y):
        y_in = fit_y[in_tr]
        j = 0
        for c, smooths in plan:
            cc = codes_tr[c][tr_idx]
            for arr in _encode_codes(cc[in_tr], y_in, ncats[c], smooths, float(y_in.mean()), float(y_in.var())):
                te_fit[in_va, j] = arr[cc[in_va]]
                j += 1
    te_va = np.zeros((len(va_idx), len(names)), dtype="float32")
    te_te = np.zeros((len(test), len(names)), dtype="float32")
    j = 0
    for c, smooths in plan:
        for arr in _encode_codes(codes_tr[c][tr_idx], fit_y, ncats[c], smooths,
                                 float(fit_y.mean()), float(fit_y.var())):
            te_va[:, j] = arr[codes_tr[c][va_idx]]
            te_te[:, j] = arr[codes_te[c]]
            j += 1
    return [pd.DataFrame(a, columns=names) for a in (te_fit, te_va, te_te)]


t0 = time.time()
xte_tr, xte_va, xte_te = target_encode_xgb(XGB_TE_PLAN)

print(f"TE {xte_tr.shape[1]} 列({time.time() - t0:.0f} 秒)")

X_tr_xgb = pd.concat([xtr.iloc[tr_idx].reset_index(drop=True), xte_tr], axis=1)
X_va_xgb = pd.concat([xtr.iloc[va_idx].reset_index(drop=True), xte_va], axis=1)
X_te_xgb = pd.concat([xte.reset_index(drop=True), xte_te], axis=1)
check_columns("xgb", X_tr_xgb.columns)

TE 29 列(2 秒)
xgb: 69 列(本番 69 列) → 列の集合 一致 / 並び順 一致


### 4-3 CatBoost — catify(値の種類（ユニーク値）が少ない数値列もカテゴリとして渡す)— +0.00170

**意図**: CatBoost は `cat_features` に指定した列に対して、内部で **Ordered Target Statistics**

(自分より前の行だけで計算する、リークしない購入率)を自動で作る。

年齢(45種)・スタンド数(15〜20種)など**値の種類（ユニーク値）が少ない数値列を文字列にして渡せば、その仕組みを数値列にも使える。**

CatBoost 単体で最大の改善(+0.00170)。

**LightGBM に横展開すると -0.00016 と逆効果だった。** 効果の源はカテゴリ化そのものではなく

CatBoost の Ordered Target Statistics にあり、LightGBM ではむしろ値の大小の順序を捨てる分だけ損をした。

このため CatBoost 専用。Count Encoding も内部の統計と重なるので使わない(3-6)。

| 部品 | LightGBM 版(3 章) | CatBoost 版 |
|---|---|---|
| Smooth Keys | 年収 /10・/100・/1000 | 年収 **/100・/1000** |
| digit | 浮動小数で計算。年収・通勤距離だけ(7 列) | XGBoost 版と同じ計算。年収・通勤距離に加えて**保有台数・環境意識の1の位**を作る(10 列)。catify で文字列にする元の列を、数値として見る唯一の列になるため。**catify の前に作る** |
| Target Encoding の平滑化 | auto / 10 / 100 | **10 / 20 / 100**。値の種類（ユニーク値）が少ない数値列は 20 だけ、カテゴリ列には作らない |
| Target Encoding のキー | 値そのもの | 値を文字列にしたもの(catify 後の列もそのまま使える) |
| Target Encoding の列名 | `te_<キー>_sauto` など | `te10_<キー>` など |

In [14]:
# 年収 /1 は年収そのものと同じキーになるので作らない
CB_SMOOTH = [("sk_inc_100", "Annual_Income_USD", 100.0), ("sk_inc_1000", "Annual_Income_USD", 1000.0),
             ("sk_com_1", "Daily_Commute_km", 1.0)]

# ① 生の 13 列 + digit(XGBoost 版と同じ関数。catify で文字列にする前に作る)
CB_DIGIT_COLS = HIGH_CARD + ["Number_of_Cars_Owned", "Environmental_Concern_Level"]   # 後の2列は 1 の位だけ
cb_d_tr, cb_d_te = digit_block(train, CB_DIGIT_COLS), digit_block(test, CB_DIGIT_COLS)
cb_digits = [c for c in cb_d_tr.columns if cb_d_tr[c].nunique() > 1 or cb_d_te[c].nunique() > 1]
cX = pd.concat([train[ALL], cb_d_tr[cb_digits]], axis=1)
cX_te = pd.concat([test[ALL], cb_d_te[cb_digits]], axis=1)
cb_features = ALL + cb_digits

# ② Smooth Keys(TE のキー専用。モデルには渡さない)
cb_sk = [name for name, _, _ in CB_SMOOTH]
cX = pd.concat([cX, smooth_key_frame(train, CB_SMOOTH)], axis=1)
cX_te = pd.concat([cX_te, smooth_key_frame(test, CB_SMOOTH)], axis=1)

# ③ catify: 値の種類（ユニーク値）が少ない数値列を文字列にして cat_features に加える
for c in LOWCARD_NUM:
    cX[c] = cX[c].astype(str)
    cX_te[c] = cX_te[c].astype(str)

cb_cats = CAT + LOWCARD_NUM

print("cat_features:", cb_cats)
print("catify 後の Age の型:", cX["Age"].dtype, "/ 例:", cX["Age"].head(3).tolist())


def target_encode_cb(tr, va, te, y_tr, plan, n_inner=5, seed=42):
    # Out-of-Fold の Target Encoding(CatBoost 版)。キーは文字列、3 つのフレームに列を直接足す
    # plan: [(キー, 平滑化の種類), ...]
    prior = float(y_tr.mean())
    inner = list(StratifiedKFold(n_splits=n_inner, shuffle=True, random_state=seed)
                 .split(np.zeros(len(tr)), y_tr))
    new = []
    for col, smooths in plan:
        k_tr, k_va, k_te = tr[col].astype(str), va[col].astype(str), te[col].astype(str)
        names = [f"te{s:g}_{col}" for s in smooths]
        full = pd.DataFrame({"k": k_tr.to_numpy(), "y": y_tr}).groupby("k")["y"].agg(["sum", "count"])
        for s, name in zip(smooths, names):
            rate = (full["sum"] + prior * s) / (full["count"] + s)
            va[name] = k_va.map(rate).astype("float64").fillna(prior)
            te[name] = k_te.map(rate).astype("float64").fillna(prior)
        oof = {name: np.full(len(tr), prior, dtype="float64") for name in names}
        for in_idx, out_idx in inner:
            agg = pd.DataFrame({"k": k_tr.iloc[in_idx].to_numpy(), "y": y_tr[in_idx]}).groupby("k")["y"].agg(["sum", "count"])
            k_out = k_tr.iloc[out_idx]
            for s, name in zip(smooths, names):
                vals = k_out.map((agg["sum"] + prior * s) / (agg["count"] + s)).astype("float64").to_numpy()
                oof[name][out_idx] = np.where(np.isnan(vals), prior, vals)
        for name in names:
            tr[name] = oof[name]
        new += names
    return new


# ④ Target Encoding: 値の種類（ユニーク値）が多い列と Smooth Keys は平滑化 3 種、値の種類（ユニーク値）が少ない数値列は 20 だけ、カテゴリ列には作らない
CB_TRIPLE = (10.0, 20.0, 100.0)
CB_TE_PLAN = ([(c, CB_TRIPLE if c in HIGH_CARD else (20.0,)) for c in NUM]
              + [(k, CB_TRIPLE) for k in cb_sk])
cb_tr, cb_va, cb_te = cX.iloc[tr_idx].copy(), cX.iloc[va_idx].copy(), cX_te.copy()
t0 = time.time()
cb_new_te = target_encode_cb(cb_tr, cb_va, cb_te, y[tr_idx], CB_TE_PLAN)

print(f"TE {len(cb_new_te)} 列({time.time() - t0:.0f} 秒)")

cb_cols = cb_features + cb_new_te
check_columns("catboost", cb_cols)

cat_features: ['Gender', 'City_Type', 'Current_Car_Type', 'Home_Charging_Possible', 'Subsidy_Available', 'Range_Anxiety_Level', 'Age', 'Number_of_Cars_Owned', 'Charging_Stations_Near_Home', 'Charging_Stations_Near_Work', 'Environmental_Concern_Level']
catify 後の Age の型: str / 例: ['66', '38', '26']


TE 20 列(7 秒)
catboost: 43 列(本番 43 列) → 列の集合 一致 / 並び順 一致


### 4-4 GBDT 3種が使う列の一覧

4-1〜4-3 で組み上げた列を、モデルごとに書き出しておく。`# ----` は列の種類、各行の右のコメントはその列の作り方(元の列には付けていない)。

> **このリストは確認用で、実際に読み込んでいるのはここではない。** 本番では `src/04_train_and_evaluate_<model>.py` が `src/03_feature_engineering_<model>.py` の関数で同じ列を組み上げ、そのまま学習に使う(このリストは読み込まない)。
> ここでは、各モデルが最終的に何列・どの列を使っているかを一目で分かるように、あえてハードコードで載せている。列を変えたときは `--dump-features` で `docs/features_<model>.json` を作り直し、このリストも書き換える。

In [15]:
lgbm_features = [
    # ---- 生の列
    "Age",
    "Annual_Income_USD",
    "Daily_Commute_km",
    "Number_of_Cars_Owned",
    "Charging_Stations_Near_Home",
    "Charging_Stations_Near_Work",
    "Environmental_Concern_Level",
    "Gender",
    "City_Type",
    "Current_Car_Type",
    "Home_Charging_Possible",
    "Subsidy_Available",
    "Range_Anxiety_Level",
    # ---- digit / 小数の分解
    "Annual_Income_USD_digit0",                                     # floor(Annual_Income_USD) % 10
    "Annual_Income_USD_digit1",                                     # floor(Annual_Income_USD / 10) % 10
    "Annual_Income_USD_digit2",                                     # floor(Annual_Income_USD / 100) % 10
    "Annual_Income_USD_digit3",                                     # floor(Annual_Income_USD / 1000) % 10
    "Daily_Commute_km_digit-1",                                     # floor(Daily_Commute_km * 10) % 10
    "Daily_Commute_km_digit0",                                      # floor(Daily_Commute_km) % 10
    "Daily_Commute_km_digit1",                                      # floor(Daily_Commute_km / 10) % 10
    # ---- Count / Frequency
    "cnt_Annual_Income_USD",                                        # Annual_Income_USD の出現回数 / 全行数(train + test)
    "cnt_Daily_Commute_km",                                         # Daily_Commute_km の出現回数 / 全行数(train + test)
    # ---- Target Encoding
    "te_Age_sauto",                                                 # Age ごとの購入率。平滑化 m=auto(値ごとに決める)(fold 内・Out-of-Fold)
    "te_Annual_Income_USD_sauto",                                   # Annual_Income_USD ごとの購入率。平滑化 m=auto(値ごとに決める)(fold 内・Out-of-Fold)
    "te_Annual_Income_USD_s10",                                     # Annual_Income_USD ごとの購入率。平滑化 m=10(fold 内・Out-of-Fold)
    "te_Annual_Income_USD_s100",                                    # Annual_Income_USD ごとの購入率。平滑化 m=100(fold 内・Out-of-Fold)
    "te_Daily_Commute_km_sauto",                                    # Daily_Commute_km ごとの購入率。平滑化 m=auto(値ごとに決める)(fold 内・Out-of-Fold)
    "te_Daily_Commute_km_s10",                                      # Daily_Commute_km ごとの購入率。平滑化 m=10(fold 内・Out-of-Fold)
    "te_Daily_Commute_km_s100",                                     # Daily_Commute_km ごとの購入率。平滑化 m=100(fold 内・Out-of-Fold)
    "te_Number_of_Cars_Owned_sauto",                                # Number_of_Cars_Owned ごとの購入率。平滑化 m=auto(値ごとに決める)(fold 内・Out-of-Fold)
    "te_Charging_Stations_Near_Home_sauto",                         # Charging_Stations_Near_Home ごとの購入率。平滑化 m=auto(値ごとに決める)(fold 内・Out-of-Fold)
    "te_Charging_Stations_Near_Work_sauto",                         # Charging_Stations_Near_Work ごとの購入率。平滑化 m=auto(値ごとに決める)(fold 内・Out-of-Fold)
    "te_Environmental_Concern_Level_sauto",                         # Environmental_Concern_Level ごとの購入率。平滑化 m=auto(値ごとに決める)(fold 内・Out-of-Fold)
    "te_sk_inc10_sauto",                                            # floor(Annual_Income_USD / 10) ごとの購入率。平滑化 m=auto(値ごとに決める)(fold 内・Out-of-Fold)
    "te_sk_inc10_s10",                                              # floor(Annual_Income_USD / 10) ごとの購入率。平滑化 m=10(fold 内・Out-of-Fold)
    "te_sk_inc10_s100",                                             # floor(Annual_Income_USD / 10) ごとの購入率。平滑化 m=100(fold 内・Out-of-Fold)
    "te_sk_inc100_sauto",                                           # floor(Annual_Income_USD / 100) ごとの購入率。平滑化 m=auto(値ごとに決める)(fold 内・Out-of-Fold)
    "te_sk_inc100_s10",                                             # floor(Annual_Income_USD / 100) ごとの購入率。平滑化 m=10(fold 内・Out-of-Fold)
    "te_sk_inc100_s100",                                            # floor(Annual_Income_USD / 100) ごとの購入率。平滑化 m=100(fold 内・Out-of-Fold)
    "te_sk_inc1000_sauto",                                          # floor(Annual_Income_USD / 1000) ごとの購入率。平滑化 m=auto(値ごとに決める)(fold 内・Out-of-Fold)
    "te_sk_inc1000_s10",                                            # floor(Annual_Income_USD / 1000) ごとの購入率。平滑化 m=10(fold 内・Out-of-Fold)
    "te_sk_inc1000_s100",                                           # floor(Annual_Income_USD / 1000) ごとの購入率。平滑化 m=100(fold 内・Out-of-Fold)
    "te_sk_commute_sauto",                                          # floor(Daily_Commute_km) ごとの購入率。平滑化 m=auto(値ごとに決める)(fold 内・Out-of-Fold)
    "te_sk_commute_s10",                                            # floor(Daily_Commute_km) ごとの購入率。平滑化 m=10(fold 内・Out-of-Fold)
    "te_sk_commute_s100",                                           # floor(Daily_Commute_km) ごとの購入率。平滑化 m=100(fold 内・Out-of-Fold)
    "te_Home_Charging_PossibleXCharging_Stations_Near_Home_sauto",  # Home_Charging_Possible × Charging_Stations_Near_Home ごとの購入率。平滑化 m=auto(値ごとに決める)(fold 内・Out-of-Fold)
]

xgb_features = [
    # ---- 生の列
    "Age",
    "Annual_Income_USD",
    "Daily_Commute_km",
    "Number_of_Cars_Owned",
    "Charging_Stations_Near_Home",
    "Charging_Stations_Near_Work",
    "Environmental_Concern_Level",
    "Gender",
    "City_Type",
    "Current_Car_Type",
    "Home_Charging_Possible",
    "Subsidy_Available",
    "Range_Anxiety_Level",
    # ---- digit / 小数の分解
    "Age_d0",                           # floor(Age) % 10
    "Age_d1",                           # floor(Age / 10) % 10
    "Annual_Income_USD_d0",             # floor(Annual_Income_USD) % 10
    "Annual_Income_USD_d1",             # floor(Annual_Income_USD / 10) % 10
    "Annual_Income_USD_d2",             # floor(Annual_Income_USD / 100) % 10
    "Annual_Income_USD_d3",             # floor(Annual_Income_USD / 1000) % 10
    "Daily_Commute_km_d-1",             # floor(Daily_Commute_km * 10) % 10
    "Daily_Commute_km_d0",              # floor(Daily_Commute_km) % 10
    "Daily_Commute_km_d1",              # floor(Daily_Commute_km / 10) % 10
    "Daily_Commute_km_d2",              # floor(Daily_Commute_km / 100) % 10
    "Charging_Stations_Near_Home_d0",   # floor(Charging_Stations_Near_Home) % 10
    "Charging_Stations_Near_Home_d1",   # floor(Charging_Stations_Near_Home / 10) % 10
    "Charging_Stations_Near_Work_d0",   # floor(Charging_Stations_Near_Work) % 10
    "Charging_Stations_Near_Work_d1",   # floor(Charging_Stations_Near_Work / 10) % 10
    # ---- Count / Frequency
    "Age_ce",                           # Age の出現回数 / 全行数(train + test)
    "Annual_Income_USD_ce",             # Annual_Income_USD の出現回数 / 全行数(train + test)
    "Daily_Commute_km_ce",              # Daily_Commute_km の出現回数 / 全行数(train + test)
    "Number_of_Cars_Owned_ce",          # Number_of_Cars_Owned の出現回数 / 全行数(train + test)
    "Charging_Stations_Near_Home_ce",   # Charging_Stations_Near_Home の出現回数 / 全行数(train + test)
    "Charging_Stations_Near_Work_ce",   # Charging_Stations_Near_Work の出現回数 / 全行数(train + test)
    "Environmental_Concern_Level_ce",   # Environmental_Concern_Level の出現回数 / 全行数(train + test)
    "Gender_ce",                        # Gender の出現回数 / 全行数(train + test)
    "City_Type_ce",                     # City_Type の出現回数 / 全行数(train + test)
    "Current_Car_Type_ce",              # Current_Car_Type の出現回数 / 全行数(train + test)
    "Home_Charging_Possible_ce",        # Home_Charging_Possible の出現回数 / 全行数(train + test)
    "Subsidy_Available_ce",             # Subsidy_Available の出現回数 / 全行数(train + test)
    "Range_Anxiety_Level_ce",           # Range_Anxiety_Level の出現回数 / 全行数(train + test)
    # ---- Target Encoding
    "inc_f100_tea",                     # floor(Annual_Income_USD / 100) ごとの購入率。平滑化 m=auto(値ごとに決める)(fold 内・Out-of-Fold)
    "inc_f100_te10",                    # floor(Annual_Income_USD / 100) ごとの購入率。平滑化 m=10(fold 内・Out-of-Fold)
    "inc_f100_te100",                   # floor(Annual_Income_USD / 100) ごとの購入率。平滑化 m=100(fold 内・Out-of-Fold)
    "inc_f1000_tea",                    # floor(Annual_Income_USD / 1000) ごとの購入率。平滑化 m=auto(値ごとに決める)(fold 内・Out-of-Fold)
    "inc_f1000_te10",                   # floor(Annual_Income_USD / 1000) ごとの購入率。平滑化 m=10(fold 内・Out-of-Fold)
    "inc_f1000_te100",                  # floor(Annual_Income_USD / 1000) ごとの購入率。平滑化 m=100(fold 内・Out-of-Fold)
    "inc_f10000_tea",                   # floor(Annual_Income_USD / 10000) ごとの購入率。平滑化 m=auto(値ごとに決める)(fold 内・Out-of-Fold)
    "inc_f10000_te10",                  # floor(Annual_Income_USD / 10000) ごとの購入率。平滑化 m=10(fold 内・Out-of-Fold)
    "inc_f10000_te100",                 # floor(Annual_Income_USD / 10000) ごとの購入率。平滑化 m=100(fold 内・Out-of-Fold)
    "commute_f1_tea",                   # floor(Daily_Commute_km) ごとの購入率。平滑化 m=auto(値ごとに決める)(fold 内・Out-of-Fold)
    "commute_f1_te10",                  # floor(Daily_Commute_km) ごとの購入率。平滑化 m=10(fold 内・Out-of-Fold)
    "commute_f1_te100",                 # floor(Daily_Commute_km) ごとの購入率。平滑化 m=100(fold 内・Out-of-Fold)
    "Age_tea",                          # Age ごとの購入率。平滑化 m=auto(値ごとに決める)(fold 内・Out-of-Fold)
    "Annual_Income_USD_tea",            # Annual_Income_USD ごとの購入率。平滑化 m=auto(値ごとに決める)(fold 内・Out-of-Fold)
    "Annual_Income_USD_te10",           # Annual_Income_USD ごとの購入率。平滑化 m=10(fold 内・Out-of-Fold)
    "Annual_Income_USD_te100",          # Annual_Income_USD ごとの購入率。平滑化 m=100(fold 内・Out-of-Fold)
    "Daily_Commute_km_tea",             # Daily_Commute_km ごとの購入率。平滑化 m=auto(値ごとに決める)(fold 内・Out-of-Fold)
    "Daily_Commute_km_te10",            # Daily_Commute_km ごとの購入率。平滑化 m=10(fold 内・Out-of-Fold)
    "Daily_Commute_km_te100",           # Daily_Commute_km ごとの購入率。平滑化 m=100(fold 内・Out-of-Fold)
    "Number_of_Cars_Owned_tea",         # Number_of_Cars_Owned ごとの購入率。平滑化 m=auto(値ごとに決める)(fold 内・Out-of-Fold)
    "Charging_Stations_Near_Home_tea",  # Charging_Stations_Near_Home ごとの購入率。平滑化 m=auto(値ごとに決める)(fold 内・Out-of-Fold)
    "Charging_Stations_Near_Work_tea",  # Charging_Stations_Near_Work ごとの購入率。平滑化 m=auto(値ごとに決める)(fold 内・Out-of-Fold)
    "Environmental_Concern_Level_tea",  # Environmental_Concern_Level ごとの購入率。平滑化 m=auto(値ごとに決める)(fold 内・Out-of-Fold)
    "Gender_tea",                       # Gender ごとの購入率。平滑化 m=auto(値ごとに決める)(fold 内・Out-of-Fold)
    "City_Type_tea",                    # City_Type ごとの購入率。平滑化 m=auto(値ごとに決める)(fold 内・Out-of-Fold)
    "Current_Car_Type_tea",             # Current_Car_Type ごとの購入率。平滑化 m=auto(値ごとに決める)(fold 内・Out-of-Fold)
    "Home_Charging_Possible_tea",       # Home_Charging_Possible ごとの購入率。平滑化 m=auto(値ごとに決める)(fold 内・Out-of-Fold)
    "Subsidy_Available_tea",            # Subsidy_Available ごとの購入率。平滑化 m=auto(値ごとに決める)(fold 内・Out-of-Fold)
    "Range_Anxiety_Level_tea",          # Range_Anxiety_Level ごとの購入率。平滑化 m=auto(値ごとに決める)(fold 内・Out-of-Fold)
]

catboost_features = [
    # ---- 生の列
    "Age",
    "Annual_Income_USD",
    "Daily_Commute_km",
    "Number_of_Cars_Owned",
    "Charging_Stations_Near_Home",
    "Charging_Stations_Near_Work",
    "Environmental_Concern_Level",
    "Gender",
    "City_Type",
    "Current_Car_Type",
    "Home_Charging_Possible",
    "Subsidy_Available",
    "Range_Anxiety_Level",
    # ---- digit / 小数の分解
    "Annual_Income_USD_d0",              # floor(Annual_Income_USD) % 10
    "Annual_Income_USD_d1",              # floor(Annual_Income_USD / 10) % 10
    "Annual_Income_USD_d2",              # floor(Annual_Income_USD / 100) % 10
    "Annual_Income_USD_d3",              # floor(Annual_Income_USD / 1000) % 10
    "Daily_Commute_km_d-1",              # floor(Daily_Commute_km * 10) % 10
    "Daily_Commute_km_d0",               # floor(Daily_Commute_km) % 10
    "Daily_Commute_km_d1",               # floor(Daily_Commute_km / 10) % 10
    "Daily_Commute_km_d2",               # floor(Daily_Commute_km / 100) % 10
    "Number_of_Cars_Owned_d0",           # floor(Number_of_Cars_Owned) % 10
    "Environmental_Concern_Level_d0",    # floor(Environmental_Concern_Level) % 10
    # ---- Target Encoding
    "te20_Age",                          # Age ごとの購入率。平滑化 m=20(fold 内・Out-of-Fold)
    "te10_Annual_Income_USD",            # Annual_Income_USD ごとの購入率。平滑化 m=10(fold 内・Out-of-Fold)
    "te20_Annual_Income_USD",            # Annual_Income_USD ごとの購入率。平滑化 m=20(fold 内・Out-of-Fold)
    "te100_Annual_Income_USD",           # Annual_Income_USD ごとの購入率。平滑化 m=100(fold 内・Out-of-Fold)
    "te10_Daily_Commute_km",             # Daily_Commute_km ごとの購入率。平滑化 m=10(fold 内・Out-of-Fold)
    "te20_Daily_Commute_km",             # Daily_Commute_km ごとの購入率。平滑化 m=20(fold 内・Out-of-Fold)
    "te100_Daily_Commute_km",            # Daily_Commute_km ごとの購入率。平滑化 m=100(fold 内・Out-of-Fold)
    "te20_Number_of_Cars_Owned",         # Number_of_Cars_Owned ごとの購入率。平滑化 m=20(fold 内・Out-of-Fold)
    "te20_Charging_Stations_Near_Home",  # Charging_Stations_Near_Home ごとの購入率。平滑化 m=20(fold 内・Out-of-Fold)
    "te20_Charging_Stations_Near_Work",  # Charging_Stations_Near_Work ごとの購入率。平滑化 m=20(fold 内・Out-of-Fold)
    "te20_Environmental_Concern_Level",  # Environmental_Concern_Level ごとの購入率。平滑化 m=20(fold 内・Out-of-Fold)
    "te10_sk_inc_100",                   # floor(Annual_Income_USD / 100) ごとの購入率。平滑化 m=10(fold 内・Out-of-Fold)
    "te20_sk_inc_100",                   # floor(Annual_Income_USD / 100) ごとの購入率。平滑化 m=20(fold 内・Out-of-Fold)
    "te100_sk_inc_100",                  # floor(Annual_Income_USD / 100) ごとの購入率。平滑化 m=100(fold 内・Out-of-Fold)
    "te10_sk_inc_1000",                  # floor(Annual_Income_USD / 1000) ごとの購入率。平滑化 m=10(fold 内・Out-of-Fold)
    "te20_sk_inc_1000",                  # floor(Annual_Income_USD / 1000) ごとの購入率。平滑化 m=20(fold 内・Out-of-Fold)
    "te100_sk_inc_1000",                 # floor(Annual_Income_USD / 1000) ごとの購入率。平滑化 m=100(fold 内・Out-of-Fold)
    "te10_sk_com_1",                     # floor(Daily_Commute_km) ごとの購入率。平滑化 m=10(fold 内・Out-of-Fold)
    "te20_sk_com_1",                     # floor(Daily_Commute_km) ごとの購入率。平滑化 m=20(fold 内・Out-of-Fold)
    "te100_sk_com_1",                    # floor(Daily_Commute_km) ごとの購入率。平滑化 m=100(fold 内・Out-of-Fold)
]

# 4-1〜4-3 で組み上げた列と一致するかを確かめる
for name, listed, built in (("LightGBM", lgbm_features, X_tr_lgbm.columns),
                            ("XGBoost", xgb_features, X_tr_xgb.columns),
                            ("CatBoost", catboost_features, cb_cols)):
    print(f"{name}: {len(listed)} 列 / 組み上げた列と {'一致' if list(built) == listed else '不一致'}")

LightGBM: 46 列 / 組み上げた列と 一致
XGBoost: 69 列 / 組み上げた列と 一致
CatBoost: 43 列 / 組み上げた列と 一致


---

## 5. RealMLP の Feature Engineering — GBDT とは別系統(ニューラルネットワーク)

**意図: アンサンブルの中で、GBDT と違う予測の間違え方をするモデルを置くこと。**

GBDT 3種は同じ厳密値 Target Encoding を中心にしているため予測が似通い、互いに足しても伸びにくい。

RealMLP(ニューラルネット)は、値ごとの違いを Target Encoding ではなく **embedding(値ごとのベクトル)** で受け取る。

> (補足)RealMLP に GBDT と同じ厳密値 Target Encoding を足すと、単体は +0.000156 伸びたが**アンサンブルへの寄与はゼロ**だった(GBDT と予測が似て多様性が減る)。単体よりアンサンブルを優先し、入れていない(9-5)。

| 節 | 何をするか | 列の渡し方 |
|---|---|---|
| 5-1 | カテゴリ列を整数コードにする | カテゴリ |
| 5-2 | 年収を粗く丸めたキー(Smooth Keys) | カテゴリ |
| 5-3 | 数値 7 列の値そのものをカテゴリにする | カテゴリ |
| 5-4 | 通勤距離の小数部、年収が 10 の倍数かのフラグ | 数値 / カテゴリ |
| 5-5 | 元データでの年収ごとの購入率 | 数値 |
| 5-6 | 年収の出現回数 | 数値 |
| 5-7 | 年収の分位ビン(5 種類の細かさ) | カテゴリ |
| 5-8 | 組み合わせのキーと、その Target Encoding(fold 内)。本番の列との突き合わせ | カテゴリ / 数値 |
| 5-9 | 使う列の一覧 | — |

列名の末尾が `_` の列はカテゴリ(embedding へ)、先頭が `_` の列は数値として渡す。

### 5-1 カテゴリ列を整数コードにする

**意図**: NN はカテゴリ列を embedding で受け取るので、文字列を整数コードにする。

コードは train に出てきた順に割り振り、test にしか出てこない値は -1 にする(このデータには存在しない)。

以降の節でも、カテゴリにする列はすべてこの関数でコードにする。

In [16]:
# 本番(04)と同じ呼び方をすると、動作に影響しない警告が出るので表示だけ止める。
#   - sklearn TargetEncoder の引数の非推奨予告(5-8)
#   - 分位ビン: 同じ年収の人が多く、幅ゼロの区切りがまとめられたという通知(5-7)
warnings.filterwarnings("ignore", category=FutureWarning)
warnings.filterwarnings("ignore", message="Bins whose width are too small.*")

mX = train.drop(columns=["id", TARGET])
mX_te = test.drop(columns=["id"])


def fit_codes(s_tr, s_te):
    # train に出てきた順に整数コードを振り、test にも同じコードを当てる(train にない値は -1)
    codes, uniques = pd.factorize(s_tr)
    mapping = {v: i for i, v in enumerate(uniques)}
    return (np.asarray(codes, dtype="int32"),
            np.asarray(pd.Series(s_te).map(mapping).fillna(-1).astype("int32"), dtype="int32"))


for c in CAT:
    mX[c], mX_te[c] = fit_codes(mX[c], mX_te[c])

print(mX[CAT].head(3).to_string())

   Gender  City_Type  Current_Car_Type  Home_Charging_Possible  Subsidy_Available  Range_Anxiety_Level
0       0          0                 0                       0                  0                    0
1       0          1                 1                       0                  0                    0
2       1          2                 0                       1                  1                    0


### 5-2 Smooth Keys

**意図**: 3-2 の Smooth Keys と同じ。行数の少ない年収の値を、粗い解像度でも見せる。

GBDT では Target Encoding のキーにしたが、ここでは**カテゴリとしてそのまま embedding に渡す**。

刻みは年収 /100・/1000・/10000。

> 以前は通勤距離 ÷ 年齢の比(9-2)と通勤距離 /5 のキー(9-5)も渡していたが、外しても精度が変わらなかったため削除した。
> 通勤距離は1値あたり 576 行あり、年収(8 行)と違って粗くして補う必要がない。

In [17]:
for df in (mX, mX_te):
    df["Income_/_100_floor_"] = np.floor(df["Annual_Income_USD"] / 100.0).astype("int64")
    df["Income_/_1000_floor_"] = np.floor(df["Annual_Income_USD"] / 1000.0).astype("int64")
    df["Income_/_10000_floor_"] = np.floor(df["Annual_Income_USD"] / 10000.0).astype("int64")

sk_cols = ["Income_/_100_floor_", "Income_/_1000_floor_", "Income_/_10000_floor_"]

print(mX[["Annual_Income_USD", "Daily_Commute_km"] + sk_cols].head(3).to_string())
print()
print("種類数:", {c: int(mX[c].nunique()) for c in sk_cols})

   Annual_Income_USD  Daily_Commute_km  Income_/_100_floor_  Income_/_1000_floor_  Income_/_10000_floor_
0            92887.0              23.4                  928                    92                      9
1            30000.0               5.0                  300                    30                      3
2            94389.0              36.8                  943                    94                      9

種類数: {'Income_/_100_floor_': 1142, 'Income_/_1000_floor_': 150, 'Income_/_10000_floor_': 16}


### 5-3 数値 7 列の値そのものをカテゴリにする

**意図**: 3-1 の厳密値キーの NN 版。年収 13,214 種類のそれぞれに別の embedding を持たせ、

**値ごとの違いを Target Encoding を通さずに直接学ばせる。**

元の数値列もそのまま数値として残しているので、NN は「値の大小」と「値ごとの違い」の両方を使える。

In [18]:
for c in NUM:
    mX[f"{c}_cat_"], mX_te[f"{c}_cat_"] = fit_codes(np.floor(mX[c]), np.floor(mX_te[c]))

print("種類数:", {f"{c}_cat_": int(mX[f"{c}_cat_"].nunique()) for c in NUM})

種類数: {'Age_cat_': 45, 'Annual_Income_USD_cat_': 13214, 'Daily_Commute_km_cat_': 89, 'Number_of_Cars_Owned_cat_': 4, 'Charging_Stations_Near_Home_cat_': 15, 'Charging_Stations_Near_Work_cat_': 20, 'Environmental_Concern_Level_cat_': 5}


### 5-4 値の細部のフラグ — 通勤距離の小数部、年収が 10 の倍数か

**意図**: 値の細部を、単独の列として取り出して渡す。

通勤距離の小数部(0.0〜0.9)は数値、年収が 10 の倍数か(切りのいい金額か)はカテゴリとして渡す。

In [19]:
for df in (mX, mX_te):
    df["_Daily_Commute_km_decimal"] = (df["Daily_Commute_km"] % 1).round(2).astype("float32")
    df["Annual_Income_USD_is_multiple_10_"] = (np.floor(df["Annual_Income_USD"]) % 10 == 0).astype("int64")

flag = mX["Annual_Income_USD_is_multiple_10_"] == 1

print(mX[["Daily_Commute_km", "_Daily_Commute_km_decimal", "Annual_Income_USD", "Annual_Income_USD_is_multiple_10_"]].head(3).to_string())
print()
print(f"年収が 10 の倍数の行: {flag.mean():.1%}(うち最小値 30,000 ドルの行: {(mX.loc[flag, 'Annual_Income_USD'] == 30000).mean():.1%})")

   Daily_Commute_km  _Daily_Commute_km_decimal  Annual_Income_USD  Annual_Income_USD_is_multiple_10_
0              23.4                        0.4            92887.0                                  0
1               5.0                        0.0            30000.0                                  1
2              36.8                        0.8            94389.0                                  0

年収が 10 の倍数の行: 19.9%(うち最小値 30,000 ドルの行: 46.2%)


年収が 10 の倍数の行のうち、約半分(46%)は最小値 30,000 ドルの行だった。

30,000 ドルの行を除くと、10 の倍数かどうかで購入率はほとんど変わらない(18.9% 対 18.8%)。

このフラグが持つ情報は、実質的に「年収が下限値か」に近い。

### 5-5 元データでの年収ごとの購入率

**意図**: コンペのデータの元になった実データ(1万行)で、年収ごとの購入率を計算して渡す。

**コンペの train・test とは別のデータから作るので、リークしない。** アンサンブルへの寄与は +0.000022。

元データにない年収は、元データ全体の購入率で埋める。

In [20]:
orig = pd.read_csv("data/EV_Adoption_and_Range_Anxiety_Dataset.csv")
orig[TARGET] = (orig[TARGET] == "Yes").astype(int)
orig_rate = orig.groupby("Annual_Income_USD")[TARGET].mean()

for df in (mX, mX_te):
    df["_Annual_Income_USD_mean_target_orig"] = (
        df["Annual_Income_USD"].map(orig_rate).fillna(orig[TARGET].mean()).astype("float32"))

print(f"元データ {orig.shape[0]:,} 行 / train の年収が元データにもある行: "
      f"{mX['Annual_Income_USD'].isin(orig_rate.index).mean():.1%}")
print(mX[["Annual_Income_USD", "_Annual_Income_USD_mean_target_orig"]].head(3).to_string())

元データ 10,000 行 / train の年収が元データにもある行: 97.9%
   Annual_Income_USD  _Annual_Income_USD_mean_target_orig
0            92887.0                             0.000000
1            30000.0                             0.066427
2            94389.0                             1.000000


### 5-6 年収の出現回数

**意図**: 3-6 の Count Encoding と同じ。その年収が珍しい値かどうかを渡す。

GBDT 版と違い、**train だけで数えて**、test にも同じ表を当てる(train にない値は 0)。

In [21]:
income_counts = mX["Annual_Income_USD"].value_counts()

for df in (mX, mX_te):
    df["_Annual_Income_USD_count"] = df["Annual_Income_USD"].astype(object).map(income_counts).fillna(0).astype("int32")

print(mX[["Annual_Income_USD", "_Annual_Income_USD_count"]].head(3).to_string())
print()
print(f"test にしかない年収の行: {(mX_te['_Annual_Income_USD_count'] == 0).mean():.2%}")

   Annual_Income_USD  _Annual_Income_USD_count
0            92887.0                        96
1            30000.0                     61605
2            94389.0                        45

test にしかない年収の行: 0.58%


### 5-7 年収の分位ビン — 5 種類の細かさ

**意図**: 年収を、**各区間の行数がそろうように**区切る(分位ビン)。400 / 600 / 800 / 900 / 1100 分位の 5 種類を並べて渡す。

5-2 の Smooth Keys は金額で等間隔に区切るので、人の多い金額帯ほど1区間の行数が多くなる。

分位ビンは行数で区切るので、どの金額帯でも同じくらいの行数で購入率を学べる。

区切りは train で決め、test にも同じ区切りを当てる。

In [22]:
bin_cols = []

for n_bins in [400, 600, 800, 900, 1100]:
    kb = KBinsDiscretizer(n_bins=n_bins, encode="ordinal", strategy="quantile", subsample=None)
    name = f"Annual_Income_USD_{n_bins}_quantile_bin_"
    mX[name] = kb.fit_transform(mX[["Annual_Income_USD"]]).ravel().astype("int32")
    mX_te[name] = kb.transform(mX_te[["Annual_Income_USD"]]).ravel().astype("int32")
    bin_cols.append(name)

# 同じ年収の人が多い区間(最小値 30,000 ドルなど)は区切れないので、実際の区間数は指定より少なくなる
print("実際の区間数:", {c: int(mX[c].nunique()) for c in bin_cols})

実際の区間数: {'Annual_Income_USD_400_quantile_bin_': 364, 'Annual_Income_USD_600_quantile_bin_': 545, 'Annual_Income_USD_800_quantile_bin_': 726, 'Annual_Income_USD_900_quantile_bin_': 818, 'Annual_Income_USD_1100_quantile_bin_': 995}


### 5-8 組み合わせの Target Encoding と、本番の列との突き合わせ

**意図**: 年収 × 航続距離の不安、年齢 × 航続距離の不安、自宅充電の可否 × 自宅スタンド数の3組を、組み合わせのキーにして Target Encoding する。

3組目は EDA で見つかった唯一の交互作用で、RealMLP では +0.000028(z=+1.96)。採用基準には届かないが、CV が最高の最終構成として採用。

**年収そのもの(5-3 で整数コードにした列)も、同じ仕組みで Target Encoding する**(2026-09-28 採用)。5-5 の元データの購入率と並べて渡すと、単体 +0.000128(z=+6.84)/ アンサンブル +0.000018(z=+3.06)で、CV が最高になった。元データの購入率だけ(train を使わない)でも、この列だけ(元データを使わない)でも、ここまでは伸びない。

キー自体もカテゴリとして embedding に渡す。

Target Encoding は目的変数を使うので、3-4 と同じく **fold 1 の学習行だけで作る**

(sklearn の `TargetEncoder`。学習行は内側 5 分割の OOF 値、検証行・test は学習行すべてで計算した値)。

In [23]:
# 組み合わせのキー(2 列の値を文字列でつなぎ、整数コードにする)
COMBOS = [("Annual_Income_USD", "Range_Anxiety_Level"), ("Age", "Range_Anxiety_Level"),
          ("Home_Charging_Possible", "Charging_Stations_Near_Home")]
combo_names = []

for a, b in COMBOS:
    name = f"{a}_{b}_"
    mX[name], mX_te[name] = fit_codes(mX[a].astype(str) + "_" + mX[b].astype(str),
                                      mX_te[a].astype(str) + "_" + mX_te[b].astype(str))
    combo_names.append(name)

print("組み合わせキーの種類数:", {c: int(mX[c].nunique()) for c in combo_names})

# fold 内の Target Encoding
m_tr, m_va, m_te = mX.iloc[tr_idx].copy(), mX.iloc[va_idx].copy(), mX_te.copy()
enc = TargetEncoder(cv=5, smooth="auto", shuffle=True, random_state=42)
te_cols = combo_names + ["Annual_Income_USD_cat_"]   # 組み合わせ 3 組 + 年収そのもの
te_names = [f"_{c}TE" for c in te_cols]
m_tr[te_names] = enc.fit_transform(m_tr[te_cols], y[tr_idx])
m_va[te_names] = enc.transform(m_va[te_cols])
m_te[te_names] = enc.transform(m_te[te_cols])

print(m_va[te_names].describe().loc[["mean", "min", "max"]].round(4).to_string())

# 列名の末尾が "_" ならカテゴリ(embedding へ)、先頭が "_" なら数値として渡す
mlp_cats = CAT + [c for c in mX.columns if c.endswith("_")]
mlp_nums = NUM + [c for c in mX.columns if c.startswith("_")]

print(f"カテゴリとして渡す列 {len(mlp_cats)} 本 / 数値として渡す列 {len(mlp_nums)} 本")
check_columns("realmlp", m_tr.columns)

組み合わせキーの種類数: {'Annual_Income_USD_Range_Anxiety_Level_': 21543, 'Age_Range_Anxiety_Level_': 135, 'Home_Charging_Possible_Charging_Stations_Near_Home_': 30}


      _Annual_Income_USD_Range_Anxiety_Level_TE  _Age_Range_Anxiety_Level_TE  _Home_Charging_Possible_Charging_Stations_Near_Home_TE  _Annual_Income_USD_cat_TE
mean                                     0.1737                       0.1745                                                  0.1748                     0.1732
min                                      0.0000                       0.0000                                                  0.0590                     0.0000
max                                      1.0000                       0.2352                                                  0.2183                     1.0000
カテゴリとして渡す列 25 本 / 数値として渡す列 10 本
realmlp: 39 列(本番 39 列) → 列の集合 一致 / 並び順 一致


### 5-9 RealMLP が使う列の一覧

5-1〜5-8 で組み上げた列を書き出しておく。`# ----` は列の種類、各行の右のコメントはその列の作り方(元の列には付けていない)。

> 4-4 と同じく確認用の一覧で、本番はここを読み込まない。列を変えたときは `docs/features_realmlp.json` とあわせて書き換える。

In [24]:
realmlp_features = [
    # ---- 生の列
    "Age",
    "Annual_Income_USD",
    "Daily_Commute_km",
    "Number_of_Cars_Owned",
    "Charging_Stations_Near_Home",
    "Charging_Stations_Near_Work",
    "Environmental_Concern_Level",
    "Gender",
    "City_Type",
    "Current_Car_Type",
    "Home_Charging_Possible",
    "Subsidy_Available",
    "Range_Anxiety_Level",
    # ---- Smooth Keys(粗い解像度)
    "Income_/_100_floor_",                                     # floor(Annual_Income_USD / 100)(カテゴリ)
    "Income_/_1000_floor_",                                    # floor(Annual_Income_USD / 1000)(カテゴリ)
    "Income_/_10000_floor_",                                   # floor(Annual_Income_USD / 10000)(カテゴリ)
    # ---- catify(数値→カテゴリ)
    "Age_cat_",                                                # floor(Age) を train での出現順に整数コード化(カテゴリ)
    "Annual_Income_USD_cat_",                                  # floor(Annual_Income_USD) を train での出現順に整数コード化(カテゴリ)
    "Daily_Commute_km_cat_",                                   # floor(Daily_Commute_km) を train での出現順に整数コード化(カテゴリ)
    "Number_of_Cars_Owned_cat_",                               # floor(Number_of_Cars_Owned) を train での出現順に整数コード化(カテゴリ)
    "Charging_Stations_Near_Home_cat_",                        # floor(Charging_Stations_Near_Home) を train での出現順に整数コード化(カテゴリ)
    "Charging_Stations_Near_Work_cat_",                        # floor(Charging_Stations_Near_Work) を train での出現順に整数コード化(カテゴリ)
    "Environmental_Concern_Level_cat_",                        # floor(Environmental_Concern_Level) を train での出現順に整数コード化(カテゴリ)
    # ---- digit / 小数の分解
    "_Daily_Commute_km_decimal",                               # Daily_Commute_km % 1(小数部)
    # ---- フラグ
    "Annual_Income_USD_is_multiple_10_",                       # floor(Annual_Income_USD) % 10 == 0 なら 1(カテゴリ)
    # ---- 元データ由来
    "_Annual_Income_USD_mean_target_orig",                     # 元データ(1万行)での Annual_Income_USD ごとの購入率
    # ---- Count / Frequency
    "_Annual_Income_USD_count",                                # train での Annual_Income_USD の出現回数
    # ---- ビン分割
    "Annual_Income_USD_400_quantile_bin_",                     # Annual_Income_USD を 400 分位で区切った番号(train で区切りを決める。カテゴリ)
    "Annual_Income_USD_600_quantile_bin_",                     # Annual_Income_USD を 600 分位で区切った番号(train で区切りを決める。カテゴリ)
    "Annual_Income_USD_800_quantile_bin_",                     # Annual_Income_USD を 800 分位で区切った番号(train で区切りを決める。カテゴリ)
    "Annual_Income_USD_900_quantile_bin_",                     # Annual_Income_USD を 900 分位で区切った番号(train で区切りを決める。カテゴリ)
    "Annual_Income_USD_1100_quantile_bin_",                    # Annual_Income_USD を 1100 分位で区切った番号(train で区切りを決める。カテゴリ)
    # ---- 交互作用キー
    "Annual_Income_USD_Range_Anxiety_Level_",                  # Annual_Income_USD + "_" + Range_Anxiety_Level を整数コード化(カテゴリ)
    "Age_Range_Anxiety_Level_",                                # Age + "_" + Range_Anxiety_Level を整数コード化(カテゴリ)
    "Home_Charging_Possible_Charging_Stations_Near_Home_",     # Home_Charging_Possible + "_" + Charging_Stations_Near_Home を整数コード化(カテゴリ)
    # ---- Target Encoding
    "_Annual_Income_USD_Range_Anxiety_Level_TE",               # Annual_Income_USD × Range_Anxiety_Level ごとの購入率(sklearn TargetEncoder、fold 内)
    "_Age_Range_Anxiety_Level_TE",                             # Age × Range_Anxiety_Level ごとの購入率(sklearn TargetEncoder、fold 内)
    "_Home_Charging_Possible_Charging_Stations_Near_Home_TE",  # Home_Charging_Possible × Charging_Stations_Near_Home ごとの購入率(sklearn TargetEncoder、fold 内)
    "_Annual_Income_USD_cat_TE",                               # Annual_Income_USD ごとの購入率(sklearn TargetEncoder、fold 内。train だけで作る)
]

# 5-1〜5-8 で組み上げた列と一致するかを確かめる
print(f"RealMLP: {len(realmlp_features)} 列 / 組み上げた列と {'一致' if list(m_tr.columns) == realmlp_features else '不一致'}")

RealMLP: 39 列 / 組み上げた列と 一致


---

## 6. スコア(04 の実行結果のメモ)

**学習と評価は `src/04_train_and_evaluate_<model>.py` で定義・実行する。** このノートブックは学習しない。

04 は `src/03_feature_engineering_<model>.py` の関数を import し、このノートブックと同じ処理を 5-fold すべてで回したうえで学習・評価する(同じ結果になることは 7 章で確認)。

下の値は、04 が保存した OOF 予測(`oof/oof_<model>.npy`)から計算したもの(2026-09-29 時点)。

| モデル | OOF AUC | アンサンブル | 1回の所要時間(フル 5-fold) |
|---|---|---|---|
| LightGBM | 0.946200(交互作用の制約あり) | 1/3 | 約 3〜4 分 |
| XGBoost | 0.946087 | 1/3 | 約 8 分 |
| RealMLP | 0.946025 | 1/3 | 約 40〜50 分 |
| CatBoost | 0.945924 | 0(他の GBDT と同質で選ばれない) | 約 20〜25 分 |
| **アンサンブル**(順位の平均) | **0.946310** | — | 未提出(最終提出の構成は CV 0.946264 / Public 0.94642) |

> LightGBM の交互作用の制約(年収系の列とそれ以外を同じ枝で組み合わせない)は、Feature Engineering ではなくモデルの設定なので、`04_train_and_evaluate.ipynb` の 2 章に記載(2026-09-29 採用)。

> **最終提出は2本**: D + RealMLP の年収の Target Encoding の構成(CV 0.946264)と、整理前の元の構成(CV 0.946234 / Public 0.94645。Public が最高)。
> 元の構成のコードは git のタグ `final-original-20260927` に残してある

本番の実行コマンドは README の「現行ベストの再現コマンド」、または `04_train_and_evaluate.ipynb` の 6 章。

---

## 7. `.py` と同じ結果になるかの確認

ここだけ `src/` の `.py` を読み込み、本番(04)と同じ呼び方で fold 1 の行列を作って、上で作った行列と**値まで**比べる。

学習行・検証行・test のそれぞれについて、列名と並び・型・値の最大差を出す。

本番の .py もこのノートブックも、`te_plan()`(3-3 の設計)どおりに必要な列だけを作る。ただしノートブックの関数は説明のために書き直したものなので、**できあがる行列が値まで同じであることを、ここで確かめる。**

In [25]:
import sys, importlib
sys.path.insert(0, os.path.abspath("src"))
fe_lgbm, fe_xgb, fe_cb, fe_mlp = (importlib.import_module(f"03_feature_engineering_{n}")
                                  for n in ("lgbm", "xgb", "catboost", "realmlp"))


def compare(label, nb_frame, py_frame):
    # 列名と並び・型・値の最大差を比べる(カテゴリや文字列の列は文字列として一致を見る)
    a, b = nb_frame.reset_index(drop=True), py_frame.reset_index(drop=True)
    same_cols = list(a.columns) == list(b.columns)
    same_types = same_cols and all(str(a[c].dtype) == str(b[c].dtype) for c in a.columns)
    worst = 0.0
    for c in (a.columns if same_cols else []):
        if pd.api.types.is_numeric_dtype(a[c]) and pd.api.types.is_numeric_dtype(b[c]):
            worst = max(worst, float(np.max(np.abs(a[c].to_numpy("float64") - b[c].to_numpy("float64")))))
        elif not (a[c].astype(str).to_numpy() == b[c].astype(str).to_numpy()).all():
            worst = np.inf

    print(f"  {label:6s}: 列名と並び {'一致' if same_cols else '不一致'} / 型 {'一致' if same_types else '不一致'}"
          f" / 値の最大差 {worst:.1e}{'(完全一致)' if same_cols and worst == 0 else ''}")

In [26]:
# LightGBM(04_train_and_evaluate_lgbm.py の prepare / fold_matrices と同じ呼び方)
k_tr, k_te = fe_lgbm.make_key_frame(train, test)
k_tr, k_te = fe_lgbm.add_smooth_keys(k_tr, train), fe_lgbm.add_smooth_keys(k_te, test)
p_dig = fe_lgbm.add_digit_features(train, cols=fe_lgbm.HIGH_CARD_NUMERIC)
p_dig_te = fe_lgbm.add_digit_features(test, cols=fe_lgbm.HIGH_CARD_NUMERIC, keep=list(p_dig.columns))
p_cnt, p_cnt_te = fe_lgbm.count_encode(k_tr, k_te, fe_lgbm.HIGH_CARD_NUMERIC)
p_te, (p_te_va, p_te_te) = fe_lgbm.target_encode_plan(
    k_tr.iloc[tr_idx], y[tr_idx], [k_tr.iloc[va_idx], k_te], fe_lgbm.te_plan(), n_inner=5, seed=42)
p_trc, p_tec = fe_lgbm.make_categorical(train, test)
S = pd.concat([p_trc[fe_lgbm.NUMERIC_COLS + fe_lgbm.CATEGORICAL_COLS], p_dig, p_cnt], axis=1)
S_te = pd.concat([p_tec[fe_lgbm.NUMERIC_COLS + fe_lgbm.CATEGORICAL_COLS], p_dig_te, p_cnt_te], axis=1)

print("LightGBM")
compare("学習行", X_tr_lgbm, pd.concat([S.iloc[tr_idx].reset_index(drop=True), p_te.reset_index(drop=True)], axis=1))
compare("検証行", X_va_lgbm, pd.concat([S.iloc[va_idx].reset_index(drop=True), p_te_va.reset_index(drop=True)], axis=1))
compare("test", X_te_lgbm, pd.concat([S_te.reset_index(drop=True), p_te_te.reset_index(drop=True)], axis=1))

LightGBM


  学習行   : 列名と並び 一致 / 型 一致 / 値の最大差 0.0e+00(完全一致)


  検証行   : 列名と並び 一致 / 型 一致 / 値の最大差 0.0e+00(完全一致)


  test  : 列名と並び 一致 / 型 一致 / 値の最大差 0.0e+00(完全一致)


In [27]:
# XGBoost(04_train_and_evaluate_xgb.py の prepare / fold_matrices と同じ呼び方)
A = fe_xgb.ALL_COLS
q_tr, q_te = train[A].copy(), test[A].copy()
q_tr, q_te = fe_xgb.add_digit_features(q_tr, q_te, train, test, cols=fe_xgb.DIGIT_COLS)
q_tr, q_te = fe_xgb.add_count_encoding(q_tr, q_te, train, test, A)
q_tr, q_te = fe_xgb.as_ordinal(q_tr, q_te, fe_xgb.CATEGORICAL_COLS)
q_sk_tr, q_sk_te = fe_xgb.make_smooth_keys(train, test)
q_plan = fe_xgb.te_plan(list(q_sk_tr.columns))
q_c_tr, q_c_te, q_n = fe_xgb.prepare_te_codes(pd.concat([q_sk_tr, train[A]], axis=1),
                                               pd.concat([q_sk_te, test[A]], axis=1), [c for c, _ in q_plan])
q_te_tr, q_te_va, q_te_te = fe_xgb.fit_apply_te_cv_nested_plan(q_c_tr, q_c_te, q_n, pd.Series(y), q_plan, tr_idx, va_idx)

print("XGBoost")
compare("学習行", X_tr_xgb, pd.concat([q_tr.iloc[tr_idx].reset_index(drop=True), q_te_tr], axis=1))
compare("検証行", X_va_xgb, pd.concat([q_tr.iloc[va_idx].reset_index(drop=True), q_te_va], axis=1))
compare("test", X_te_xgb, pd.concat([q_te.reset_index(drop=True), q_te_te], axis=1))

XGBoost


  学習行   : 列名と並び 一致 / 型 一致 / 値の最大差 0.0e+00(完全一致)


  検証行   : 列名と並び 一致 / 型 一致 / 値の最大差 0.0e+00(完全一致)


  test  : 列名と並び 一致 / 型 一致 / 値の最大差 0.0e+00(完全一致)


In [28]:
# CatBoost(04_train_and_evaluate_catboost.py の prepare / fold_matrices と同じ呼び方)
B = fe_cb.NUMERIC_COLS + fe_cb.CATEGORICAL_COLS
r_X, r_X_te = train[B].copy(), test[B].copy()
r_dig = fe_cb.add_digits(r_X, cols=fe_cb.DIGIT_COLS)
fe_cb.add_digits(r_X_te, cols=fe_cb.DIGIT_COLS)
r_feats = B + fe_cb.drop_constant([r_X, r_X_te], r_dig)
r_sk = fe_cb.add_smooth_keys(r_X, fe_cb.PROD_SMOOTH_KEY_SPECS)
fe_cb.add_smooth_keys(r_X_te, fe_cb.PROD_SMOOTH_KEY_SPECS)
fe_cb.cast_to_str([r_X, r_X_te], fe_cb.LOWCARD_NUM_COLS)
r_tr, r_va, r_te = r_X.iloc[tr_idx].copy(), r_X.iloc[va_idx].copy(), r_X_te.copy()
r_new = fe_cb.target_encode_plan(r_tr, r_va, r_te, y[tr_idx], fe_cb.te_plan(r_sk))
r_cols = r_feats + r_new

print("CatBoost")
compare("学習行", cb_tr[cb_cols], r_tr[r_cols])
compare("検証行", cb_va[cb_cols], r_va[r_cols])
compare("test", cb_te[cb_cols], r_te[r_cols])

CatBoost


  学習行   : 列名と並び 一致 / 型 一致 / 値の最大差 0.0e+00(完全一致)


  検証行   : 列名と並び 一致 / 型 一致 / 値の最大差 0.0e+00(完全一致)


  test  : 列名と並び 一致 / 型 一致 / 値の最大差 0.0e+00(完全一致)


In [29]:
# RealMLP(04_train_and_evaluate_realmlp.py と同じ呼び方)
p_orig = fe_mlp.load_orig("data/EV_Adoption_and_Range_Anxiety_Dataset.csv")
s_X = train.drop(columns=["id", TARGET])
s_X_te = test.drop(columns=["id"])
s_map = {}
home = [("Home_Charging_Possible", "Charging_Stations_Near_Home")]   # 04 の --combo-home と同じ
s_X, s_newcat, s_newnum, s_combos = fe_mlp.build_features(s_X, CAT, NUM, s_map, fit=True, orig=p_orig, extra_combos=home)
s_X_te, _, _, _ = fe_mlp.build_features(s_X_te, CAT, NUM, s_map, fit=False, orig=p_orig, extra_combos=home)
s_tr, s_va, s_te = s_X.iloc[tr_idx].copy(), s_X.iloc[va_idx].copy(), s_X_te.copy()
s_enc = TargetEncoder(cv=5, smooth="auto", shuffle=True, random_state=42)
s_cols = s_combos + ["Annual_Income_USD_cat_"]   # 04 の --te-income と同じ
s_names = [f"_{c}TE" for c in s_cols]
s_tr[s_names] = s_enc.fit_transform(s_tr[s_cols], pd.Series(y[tr_idx]))
s_va[s_names] = s_enc.transform(s_va[s_cols])
s_te[s_names] = s_enc.transform(s_te[s_cols])

print("RealMLP")
print(f"  カテゴリ・数値の振り分け: {'一致' if (mlp_cats == CAT + s_newcat and mlp_nums == NUM + s_newnum) else '不一致'}")
compare("学習行", m_tr, s_tr)
compare("検証行", m_va, s_va)
compare("test", m_te, s_te)

RealMLP
  カテゴリ・数値の振り分け: 一致


  学習行   : 列名と並び 一致 / 型 一致 / 値の最大差 0.0e+00(完全一致)
  検証行   : 列名と並び 一致 / 型 一致 / 値の最大差 0.0e+00(完全一致)


  test  : 列名と並び 一致 / 型 一致 / 値の最大差 0.0e+00(完全一致)


### 採否表と本番の列の突き合わせ

`src/feature_catalog.py` の `FUNC_STATUS`(〇 採用 / ✖ 不採用)が、`docs/features_*.json` の列と矛盾していないかを確かめる。

In [30]:
import feature_catalog as fc
ng = fc.verify_status()

print("採否表と本番の列の矛盾:", ng if ng else "なし")

採否表と本番の列の矛盾: なし


**コードの実行はここまで。** 以降の 8・9 章は説明だけで、実行するコードはない。

---

## 8. (補足)`.py` との構成の違い

| 観点 | `.py`(`src/03_*.py` / `src/04_*.py`) | このノートブック |
|---|---|---|
| 分け方 | **モデルごと**にファイルを分ける | **GBDT 共通 → GBDT モデル別 → RealMLP** の順に並べる |
| コードの置き場所 | `03` に関数を定義し、`04` が import して 5-fold で学習する。不採用の関数は `03_feature_engineering_all.py` に記録として移してある(`FUNC_STATUS`) | ノートブックの中に採用した処理だけを書き、fold 1 で実行して確かめる(学習はしない) |
| 作る列 | `te_plan()` どおりに必要な列だけを作る | 同じ(3-3)。関数は説明のために書き直している |
| 本番との一致の確認 | `--dump-features` で列名を `docs/features_*.json` に書き出す | その json と列を突き合わせ(4・5章)、値まで一致するかを 7 章で確かめる |

---

## 9. (補足)不採用となった Feature Engineering 

検証したうえで本番に入れなかった施策。**コードは再検証を防ぐ記録として `src/03_feature_engineering_all.py` の「不採用(記録)」に残してある**(`FUNC_STATUS` で ✖)。各モデルのファイルから移した関数は、名前の末尾にもとのモデル名が付いている(例: `add_arithmetic_xgb`)。表の「ファイル」は、もとのモデル。

「変化」は OOF AUC の差(+ が改善)。2026-09-29 までに検証したものをすべて載せている。

| 節 | 内容 |
|---|---|
| 9-1 | EDA の仮説から検証したもの |
| 9-2 | 四則演算 |
| 9-3 | 交互作用の Target Encoding・行フィンガープリント |
| 9-4 | エンコード方式・Count Encoding・catify の横展開 |
| 9-5 | RealMLP への GBDT 系の情報の追加 |
| 9-6 | 外部データ(元データ)の使い方 |
| 9-7 | Feature Importance から立てた仮説(2026-09-29) |

### 9-1 EDA の仮説から検証したもの

#### (1) 列の削除: `City_Type`(仮説4)

**仮説**: `City_Type` はスタンド数からほぼ復元できる(自宅スタンド 8 個以上の 150,407 行のうち、都市居住でないのは 15 行)。

情報が重複しているので、落としても損をしないのではないか。

| 削除した列 | 列数 | 変化 | z | 判定 |
|---|---|---|---|---|
| `City_Type` | 92 → 87 | -0.000017 | -1.30 | 誤差 |

> 情報が重複していたのは事実だが、削除しても精度は向上しない。-> 削除しない。

#### (2) 列の削除: `Age` / `Gender` / `Number_of_Cars_Owned`(仮説1・3)

**仮説**: 3列とも、目的変数との関係が弱く、他の列ともほぼ無関係だった。

ノイズの可能性が高く、落とせば1本の木が有用な列を引く確率が上がるのではないか。

| 落とした列 | 列数 | 変化 | z | 判定 |
|---|---|---|---|---|
| 3列 | 92 → 74 | **-0.000441** | **-14.19** | **明確に悪化** |
| 3列(1本の木が見る列の数をそろえて再測定) | 92 → 74 | -0.000432 | -13.99 | 悪化(副作用ではない) |
| 3列 + `City_Type` | 92 → 69 | -0.000425 | -13.70 | 悪化 |

> 目的変数との全体の関係が弱いことと、モデルにとって不要なことは別物であり、あえて削除することはしない。

#### (3) 補助金との組み合わせ(仮説5)

**仮説**: 補助金がない人は、環境意識・収入・航続距離の不安のどれを動かしても購入率が0%付近に張り付く。

補助金との組み合わせを明示すれば精度向上に効果があるのではないか。

| 作り方 | 関数・パターン | 列数 | 変化 | z | 判定 |
|---|---|---|---|---|---|
| 組み合わせをキーにした Target Encoding | `te2subsidy` | 92 → 101 | -0.000016 | -0.94 | 誤差 |
| 補助金(0/1)× 各列の積 | `add_subsidy_products` | 92 → 95 | +0.000009 | +0.65 | 誤差 |

**「床に張り付く」のは確率の見た目で、log-odds では足し算だった。**

補助金の効果は log-odds で見るとどの水準でも +4.06〜4.49 でほぼ一定。

木は `Subsidy_Available` で1回分割すれば同じ構造を表現できる。

組み合わせ列は importance 13.6% とよく使われたが、単体の補助金列の出番が減っただけで AUC は動かなかった。

列を絞った後の構成で測り直すと -0.000051(z=-2.69)と、むしろ悪化傾向だった(2026-09-27)。

2026-09-28 に RealMLP(組み合わせキー 3 組 + その Target Encoding)でも測り直したが、単体 +0.000011(z=+0.60)/ アンサンブル -0.000001 で誤差だった。条件分岐が苦手な NN でも効かない。

#### (4) NN(RealMLP)だけがGBDT系モデルでは不採用だった説明変数の組み合わせが効くのでは

**仮説**: 木は条件分岐が得意だが、NN は苦手かもしれない。GBDT には不要でも、RealMLP には効くのではないか。

**検証**: EDA で見つけた構造ごとに行を分け、各モデルの予測を観測の購入率に合わせて補正したとき、

AUC が上がるかを測った(上がれば、その構造を取りこぼしている)。単位は 0.00001、採用基準は +8。

| 構造 | LightGBM | XGBoost | CatBoost | RealMLP |
|---|---|---|---|---|
| 自宅充電 × 自宅スタンド数 | -1.1 | -1.1 | -0.9 | -0.9 |
| 都市区分 × 自宅スタンド数 | -1.1 | -1.1 | -1.1 | -1.1 |
| 補助金 × 環境意識 | -1.5 | -1.4 | -1.0 | -4.1 |
| 補助金 × 収入 | -1.2 | -1.3 | -1.1 | -4.9 |
| 対照(意味のないランダム 30 群) | -0.1 | -0.1 | -0.1 | -0.1 |

> すべての構造・すべてのモデルで、補正すると AUC 下がった。RealMLP でも説明変数単体で判定できている。

### 9-2 四則演算系の特徴量作成 — パターン別

EDA とは別に、Feature Engineering として一通り試した。

**試したパターン**

| パターン | 作った列(例) | 狙い |
|---|---|---|
| 和(sum) | 自宅 + 職場のスタンド数 / 数値列の全ペアの和 | 充電環境の総量のような「合計」に意味があるか |
| 差(diff) | 自宅 − 職場のスタンド数 / 数値列の全ペアの差 | 2つの量のバランスに意味があるか |
| 比(ratio) | 年収 ÷ 通勤距離、年収 ÷ 年齢、年収 ÷ 保有台数、通勤距離 ÷ スタンド数、自宅 ÷ 職場スタンド数、通勤距離 ÷ 年齢(RealMLP) | 「1台あたり」「1 km あたり」のような単位あたりの量 |
| 積(product) | 年収 × 環境意識、年齢 × 環境意識、通勤距離 × 年齢、環境意識 × スタンド数 | 2つの効果が掛け算で効くか |
| 平均(avg) | (自宅 + 職場) ÷ 2 / 7ペアの平均 | 和と同じ(木にとっては和を 2 で割っても区別がつかない) |
| 行方向の集約 | 数値列を標準化した平均・標準偏差 / スタンド数の最小・最大・合計 | 行全体の「高さ」「ばらつき」 |

**測定結果**

| 関数 | ファイル | 含むパターン | 列数 | 変化 | 判定 |
|---|---|---|---|---|---|
| `add_arithmetic_meaningful` | lgbm | 和・差・比・積(意味で選んだ12列) | 12 | -0.00014 → 本番構成で再測定 -0.000048(z=-2.58) | 誤差 |
| `add_arithmetic_all_pairs` | lgbm | 和・差・比 × 数値7列の全21ペア | 63 | +0.00030(3-fold)→ 再測定 **-0.000089(z=-4.99)** | **悪化** |
| `add_arithmetic` / `all_numeric_pairs` | xgb | 和・差・比・平均 × 意味で選んだ7ペア | 28 | +0.00004 → 再測定 -0.000049(z=-3.22) | 誤差 |
| `add_arithmetic` | catboost | 和・差・比・積・平均(14列) | 14 | **-0.00099** | **悪化**(最大) |
| `arithmetic_meaningful` | all | 上の集約版 | — | — | 同上 |
| `add_group_means` | lgbm | 行方向の平均・標準偏差 | 3 | 効果なし | 不採用 |
| `add_row_aggregates` | xgb | スタンド数の最小・最大・合計 | 3 | 効果なし | 不採用 |
| `build_features(ratio=True)` | realmlp | 比(通勤距離 ÷ 年齢) | 1 | 外しても単体 -0.000011(z=-0.75)/ アンサンブル -0.000004(z=-0.88) | 誤差 → 2026-09-27 に削除 |

> **4モデルとも、どのパターンでも効果なし。** EDA のとおり各列の効果は log-odds 上の足し算で、「比」や「積」の関係は生成過程に入っていないと考えられる。

> RealMLP の「通勤距離 ÷ 年齢」は、公開カーネルの前処理をまとめて移植したときに入っていたもので、個別には検証していなかった。
> 2026-09-27 に外して測ったところ差は誤差だった。比は通勤距離とほぼ同じ順位(順位相関 0.885)で、単体の AUC は通勤距離そのものより低い(0.526 < 0.534)。割る相手の年齢に購入との関係がほとんどないため、通勤距離に年齢のノイズを混ぜただけの列になっていた。

### 9-3 交互作用 Target Encoding・行フィンガープリント(EDA 以前に試したもの)

| 関数 | ファイル | 内容 | 変化 | 理由 |
|---|---|---|---|---|
| `pair_keys` | lgbm | カテゴリ6列の全15ペアで Target Encoding | -0.00007 | 単列の Target Encoding と木の分割で取り切れている |
| `triple_keys` | lgbm | 3列の組み合わせ5組で Target Encoding(自宅充電 × 自宅スタンド数を含む組もある) | -0.00004 | 2列 → 3列 → 6列 → 13列すべて無効 |
| `make_interaction_keys` / `cat_pairs` | xgb | 2列の組み合わせキー | 無効 | 同上 |
| `add_interactions` | catboost | カテゴリの組み合わせ(10組) | -0.00008 | CatBoost は内部でも組み合わせの統計を作るため冗長 |
| `interaction_keys` / `cat_pairs` | all | 上の集約版 | — | 同上 |
| `te2catnum`(パターン) | lgbm | カテゴリ6列 × 値の種類が少ない数値5列の 30 組 | 実行せず | 3列版(`triple_keys`)が無効で、効果が見込めないため |
| `all_columns_key` | lgbm | 全13列の組み合わせ(行フィンガープリント) | 実行前に打ち切り | train の全行がユニークで原理的に機能しない(前コンペ S6E8 では最大の改善要因だった) |
| `fingerprint_key` | lgbm | 列の部分集合での組み合わせ | -0.00002 | 多列の組み合わせ自体に追加情報がない |

### 9-4 エンコード方式・Count Encoding・catify の横展開

| 関数 | ファイル | 内容 | 変化 | 理由 |
|---|---|---|---|---|
| `as_native_category` | xgb | カテゴリ列を category 型で渡す | 同点 | ordinal と差がなく、LightGBM との違いを出すため ordinal を採用(4-2) |
| `as_onehot` | xgb | One-Hot Encoding | +0.00005 → 再測定 -0.000002 | 列が増えるだけ。カテゴリは 2〜4 種類しかない |
| `ordcat`(パターン) | lgbm | カテゴリ列を整数コード(ordinal)で渡す | -0.000005(z=-0.42) | native category と予測がほぼ同じ(順位相関 0.9994)。カテゴリの情報は Target Encoding でも渡っている(4-1) |
| `add_count_encoding` | catboost | Count Encoding | -0.00017(他の列への上積みでも +0.00002) | CatBoost 内部の統計と重複(3-6) |
| `add_count_encoding`(年収・通勤距離だけ) | xgb | Count を値の種類が多い2列だけに作る | +0.00015(全13列なら +0.00049) | XGBoost は全13列で作る(4-2) |
| Smooth Keys への Count 拡張(パターン) | lgbm | Smooth Keys 4本にも Count を作る | -0.000012(z=-0.95) | 厳密値の Count と Target Encoding で足りている |
| `catify`(横展開) | lgbm / xgb | 値の種類が少ない数値列をカテゴリとして渡す | LightGBM -0.00016 / XGBoost は打ち切り | 効果の源は CatBoost の Ordered Target Statistics で、カテゴリ化そのものではない。LightGBM では値の順序を捨てる分だけ損(4-3) |

### 9-5 RealMLP への GBDT 系の情報の追加 — 単体は伸びてもアンサンブルで消える

| 関数 | ファイル | 内容 | 変化 | 理由 |
|---|---|---|---|---|
| `build_te_key_frame` / `target_encode_highcard` | realmlp | 値の種類（ユニーク値）が多い2列の厳密値 Target Encoding | 単体 +0.000156 / アンサンブル ±0 | GBDT と予測が似て、多様性が減る(5章) |
| `--digits`(オプション) | realmlp | 年収の6桁・通勤距離の3桁をカテゴリとして embedding に渡す | -0.00002(fold 1) | 値そのものの embedding(5-3)と Smooth Keys で桁の情報は取り込み済み |
| `build_features(km5=True)` | realmlp | 通勤距離 /5 のキー | 外しても単体 -0.000008(z=-0.51)/ アンサンブル -0.000003(z=-0.59) | 通勤距離は1値あたり 576 行あり粗くする必要がない。1 km 刻みのキー(5-3)とも重複 |
| `add_income_neighborhood` | all | 年収 ±r ドルの範囲の購入率・傾き・曲率 | GBDT ±0.00002 / RealMLP 単体 +0.000077(z=+3.45)/ アンサンブル -0.0000019 | 同上。Smooth Keys と厳密値 Target Encoding で取り込み済み |

### 9-6 外部データ(元データ)の使い方

コンペのデータの元になった元データ(1万行、`data/EV_Adoption_and_Range_Anxiety_Dataset.csv`)の使い方。

| 使い方 | 変化 | 判定 | 理由 |
|---|---|---|---|
| 元データの行を学習データに足す | -0.00002 | 不採用 | 66.9 万行に対して 1.5% の増量にしかならず、Kaggle の生成器が分布を変えている |
| 元データの購入の生成ルール(`buy_score`)を特徴量にする | -0.00002 | 不採用 | importance 74% で1位になったが、スコアは動かない。既存の列から再現できている |
| **元データでの年収ごとの購入率を1列だけ渡す(RealMLP、5-5)** | アンサンブル +0.000022(z=+3.86) | **採用** | 行ではなく統計1列。GBDT には入れていない |

### 9-7 Feature Importance から立てた仮説(2026-09-29)

`04_train_and_evaluate.ipynb` の 7 章で、補助金と環境意識が重要度の 7〜9 割を占め、次が年収の Target Encoding と分かった。
「残りの 1〜3 割(年収の細部)」を狙って仮説を 5 つ立てた。**採用したのは LightGBM の交互作用の制約だけ**(04 の 2 章)で、残りはこの表のとおり。

| 仮説 | 対象 | 単体の差 | アンサンブルの差 | 判定 | 理由 |
|---|---|---|---|---|---|
| 元データの年収ごとの購入率を足す(`orig_income_rate_gbdt`) | LightGBM / XGBoost / CatBoost | -0.000003 / -0.000005 / -0.000012 | ±0 | 誤差 | 年収の Target Encoding と同じ情報。RealMLP で効いたのは、train の購入率と並べたときだけ |
| Target Encoding の学習行の値を、内側の分割の乱数を変えて 5 回作り平均 | LightGBM / XGBoost | -0.000071 / -0.000083 | ±0 / -0.000015 | 単体は有意に悪化 | Out-of-Fold の値の揺れを消すと、かえって悪化する。揺れが、学習時に購入率を過信させない役割をしていると考えられる |
| 元の列ごとに分ける交互作用の制約 | LightGBM | +0.000075(z=+3.09) | +0.000038 | 誤差(基準に僅かに届かない) | 年収系とそれ以外の 2 つに分ける方が良い(採用) |
| CatBoost をアンサンブルに戻す | CatBoost | — | 最大 +0.000010 | 誤差 | 本番のまま・元データの購入率・TE の平均の 3 通りとも入る余地がない |
| 年収 /100 のキーの Target Encoding(train だけで作る) | RealMLP | -0.000035 | **-0.000042(z=-6.46)** | アンサンブルが有意に悪化 | 年収の Target Encoding(1 列)と重なり、GBDT と予測が似る |

> XGBoost の交互作用の制約は、単体 +0.000089(z=+4.20)と基準を満たすが、LightGBM に入れた構成への上積みは +0.000004(z=+0.52)で、採否は未決定。